{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# INFO-F-208: Project 1\n",
    "***Name:*** ThÃ©o Verhelst  \n",
    "***Matricule: *** 000400807  \n",
    "***Date: *** 10/26/2016  \n",
    "\n",
    "## Background and objectives\n",
    "Modern biological data is typically composed by large sequences of nucleotides or amino acids. When one wants to extract information from sequences of more than a thousand nucleotides, it becomes clear that computer-assisted methods are needed.\n",
    "\n",
    "Sequence alignment is a way to find similarities between two or more sequences. This project is an implementation of two sequence alignment algorithms, mainly intended to align amino acids sequences. This can be useful to find domains in raw protein data.\n",
    "\n",
    "This script can find global or local alignment, and local sub-alignments as well. It uses an affine gap penalty, so a linear gap penalty can be achieved by setting the open gap penalty to `0`.\n",
    "\n",
    "By default, all best solutions are displayed. However, with large sequences the number of solution can also be large (a few hundred for sequences of 800 - 1000 elements). Thus there is an option to show only the first best solution.\n",
    "\n",
    "Another objective is to have a solution display similar to the one used by the LALIGN server, so that biologists used to use LALIGN can use this tool seamlessly.\n",
    "\n",
    "## Code\n",
    "\n",
    "The full code is also available in the file `project1.py`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from enum import Enum\n",
    "from copy import deepcopy\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We define the `AminoAcid` type by using the `Enum` class, and we overload its string representation method."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "AminoAcid = Enum(\"AminoAcid\", \"A R N D C E Q G H I L K M F P S T W Y V B Z X\")\n",
    "# Make the str() function print only the amino acid letter by redefining __str__\n",
    "AminoAcid.__str__ = lambda self: self.name[-1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part 1.1\n",
    "### `Sequence` class\n",
    "The `Sequence` class uses internally a list of `Sequence` objects, and defines specials methods in order to meet the requirements specified by the statement (part 1.1 of the statement)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "class Sequence:\n",
    "    \"\"\"Represents a sequence of amino acids.\n",
    "    It is suited to compute the best alignment between two sequences.\"\"\"\n",
    "    \n",
    "    def __init__(self, filename, sequence_id):\n",
    "        \"\"\"Constructor. Loads the sequence from a file.\n",
    "        \n",
    "        Parameters:\n",
    "            filename: the filename of the file containing the sequence.\n",
    "            sequence_id: the identifier of the sequence in the file.\n",
    "                This is the string between the two first bars \"|\" on the line\n",
    "                preceding the sequence.\n",
    "            \"\"\"\n",
    "        self.load_from_file(filename, sequence_id)\n",
    "    \n",
    "    def __getitem__(self, index):\n",
    "        \"\"\"Returns the elements at the specified index. Allows the syntax\n",
    "        sequence[i].\n",
    "        \"\"\"\n",
    "        return self.sequence[index]\n",
    "        \n",
    "    def __len__(self):\n",
    "        \"\"\"Returns the lenght of the sequence. Allows the syntax len(sequence).\"\"\"\n",
    "        return len(self.sequence)\n",
    "\n",
    "    def load_from_file(self, filename, sequence_id):\n",
    "        \"\"\"Loads the sequence data from a file.\n",
    "        \n",
    "        Parameters:\n",
    "            filename: the filename of the file containing the sequence.\n",
    "            sequence_id: the identifier of the sequence in the file.\n",
    "                This is the string between the two first bars \"|\" on the line\n",
    "                preceding the sequence.\n",
    "        \"\"\"\n",
    "        found_sequence = False\n",
    "        sequence = \"\"\n",
    "        with open(filename) as file:\n",
    "            for line in file:\n",
    "                if not found_sequence:\n",
    "                    # We found the sequence if this is the right identifier line\n",
    "                    found_sequence = (line[0] == \">\"\n",
    "                            and line.split(\"|\")[1] == sequence_id)\n",
    "                else:\n",
    "                    if line[0] != \">\":\n",
    "                        sequence += line\n",
    "                    else:\n",
    "                        # Stop adding lines to the sequence\n",
    "                        found_sequence = False\n",
    "                        \n",
    "        if sequence == \"\":\n",
    "            raise RuntimeError(\"Sequence not found in the specified file.\")\n",
    "\n",
    "        try:\n",
    "            sequence = [AminoAcid[character] for character in sequence \\\n",
    "                    if not character.isspace()]\n",
    "        except KeyError as e:\n",
    "            raise RuntimeError(\"An amino acid is not valid in the sequence: \"\n",
    "                    + str(e))\n",
    "\n",
    "        # We didn't worked on self.sequence directly to preserve the internal\n",
    "        # state in case of failure\n",
    "        self.sequence = sequence"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Special methods ( `__getitem__` and `__len__`)\n",
    "These are the methods that fulfill the requirements of the statement,  as explained above."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### The loading function (`load_from_file`)\n",
    "I put the function that loads the sequence in the class `Sequence` so that the internal state of the object needs not to be given to the constructor, only the filename and the identifier of the sequence is given to the constructor.\n",
    "\n",
    "We consider that a sequence file consists of identifier lines, each of them being followed by a sequence. The sequence may span on mutliple lines. An identifier line is composed by the character \">\", then the identifier string of the sequence, then the character \"|\", then whatever.\n",
    "\n",
    "This format is the format used in the .fasta files."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Matrix class\n",
    "I created a Matrix class in order to have a level of abstraction for manipulating matrices. This allows me to choose a more efficient implementation than the Python's list in the future."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "class Matrix:\n",
    "    \"\"\"Represents a m*n matrix.\n",
    "    It is used to abstract the implementation of matrices, and allow possible\n",
    "    optimisations easily, without having to change the whole code.\n",
    "    All lines don't have to have the same length, so it may be used to represent\n",
    "    a triangular matrix.\n",
    "    \n",
    "    \"\"\"\n",
    "    def __init__(self, m = 0, n = 0, value=None):\n",
    "        \"\"\"Construct the matrix\n",
    "        \n",
    "        Parameters:\n",
    "            m, n: height and width of the matrix\n",
    "            value: the initial value to put in the cells.\n",
    "        \"\"\"\n",
    "        self.matrix = [[value] * n for i in range(m)]\n",
    "    \n",
    "    def __getitem__(self, key):\n",
    "        if isinstance(key, tuple):\n",
    "            return self.matrix[key[0]][key[1]]\n",
    "        else:\n",
    "            return self.matrix[key]\n",
    "            \n",
    "    def __setitem__(self, key, value):\n",
    "        if isinstance(key, tuple):\n",
    "            self.matrix[key[0]][key[1]] = value\n",
    "        else:\n",
    "            self.matrix[key] = value\n",
    "    \n",
    "    def __str__(self):\n",
    "        cell_width = max([max([len(str(c)) for c in line]) for line in self.matrix])\n",
    "        res = \"\"\n",
    "        for line in self.matrix:\n",
    "            for cell in line:\n",
    "                res += str(cell).rjust(cell_width) + \" \"\n",
    "            res += \"\\n\"\n",
    "        return res\n",
    "            \n",
    "    def append(self, iterable):\n",
    "        \"\"\"Adds a line to the bottom of the matrix.\"\"\"\n",
    "        self.matrix.append(list(iterable))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The methods of this class are rather straightforward, they just access/modify/display the matrix."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### `Score` class\n",
    "The `Score` class uses internally a triangular matrix (a list of lists of increasing sizes) to store the substitution data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "class Score:\n",
    "    \"\"\"Class used to compute the score of substitution between two amino acids,\n",
    "    by using well known matrix such as BLOSUM.\n",
    "    \"\"\"\n",
    "    \n",
    "    def __init__(self, filename):\n",
    "        \"\"\"Contructor. Loads the matrix from a file.\n",
    "        \n",
    "        Parameter:\n",
    "            filename: the filename of the matrix to load. It must be a .iij file\n",
    "            as given in the blosum archive.\n",
    "        \"\"\"\n",
    "        self.load_from_file(filename)\n",
    "    \n",
    "    def get_score(self, amino_acid_A, amino_acid_B):\n",
    "        \"\"\"Returns the match score between two amino acids according to\n",
    "        the loaded matrix.\n",
    "        \"\"\"\n",
    "        index_A = self.indices.index(amino_acid_A)\n",
    "        index_B = self.indices.index(amino_acid_B)\n",
    "        \n",
    "        # The matrix is triangular, so we need to test the two possibilities for\n",
    "        # indexing since one of the two would result in an index error\n",
    "        if index_A > index_B:\n",
    "            return self.matrix[index_A, index_B]\n",
    "        else:\n",
    "            return self.matrix[index_B, index_A]\n",
    "    \n",
    "    def load_from_file(self, filename):\n",
    "        \"\"\"Loads a substitution matrix from a file.\n",
    "        \n",
    "        Parameter:\n",
    "            filename: the filename of the matrix to load. It must be a .iij file\n",
    "            as given in the blosum archive.\n",
    "        \"\"\"\n",
    "        self.matrix = Matrix()\n",
    "        self.indices = []\n",
    "        \n",
    "        with open(filename) as file:\n",
    "            found_amino_acid_list = False\n",
    "            \n",
    "            for line in file:\n",
    "                # Skip commentary lines\n",
    "                if line.strip().startswith(\"#\"):\n",
    "                    continue\n",
    "                \n",
    "                if not found_amino_acid_list:\n",
    "                    # The first non-commentary line is the list of amino acids\n",
    "                    # We reload this list because some files may order the acids\n",
    "                    # differently\n",
    "                    self.indices = [AminoAcid[character] for character in line.split()]\n",
    "                    found_amino_acid_list = True\n",
    "                else:\n",
    "                    self.matrix.append([int(number) for number in line.split()])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### The score method (`get_score`)\n",
    "The method that computes the substitution score between two amino acids receives two instances of the `AminoAcid` enumeration as parameters.  \n",
    "The first thing it has to do is to convert the amino acids into indexes suitable for indexing the internal substitution matrix."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### The loading function (`load_from_file`)\n",
    "As in the `Sequence` class, the function that loads the matrix from a file is in the `Score` class itself.  \n",
    "We consider that a matrix file is composed by a line listing all amino acids reported in the matrix, and then the data of the matrix. One line of the matrix is one line of the file, and the columns are separated by spaces.  \n",
    "There may be comment lines, i.e. lines starting by the `#` character. Such lines are skipped.\n",
    "\n",
    "This format is the format used in the `*.iij` BLOSUM files linked in the statement."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part 1.2 and 1.3\n",
    "### `Aligner` class\n",
    "This class holds all the functions related to the alignment algorithm and the solutions display.\n",
    "\n",
    "I decided to make this class because many objects are used accros the multiple functions of the algorithm.\n",
    "By making a class, such objects are now attributes, and need not to be passed as parameters anymore. This reduces stack overloading when aligning very large sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "class Aligner:\n",
    "    \"\"\"Holds functions and variable related to the sequence alignment.\"\"\"\n",
    "    \n",
    "    move_values = {\"D\" : (-1, -1), \"T\" : (-1, 0), \"L\" : (0, -1), \"0\" : (0, 0)}\n",
    "\n",
    "    def __init__(self, score, sequence_A, sequence_B,\n",
    "            open_penalty = -12,\n",
    "            extend_penalty = -2,\n",
    "            local = True,\n",
    "            sub_alignments = 0,\n",
    "            display_all_solutions = True,\n",
    "            max_line_length = 80):\n",
    "        \"\"\"Constructor, with all the parameters of the algorithm.\"\"\"\n",
    "        self.score = score\n",
    "        self.sequence_A = sequence_A\n",
    "        self.sequence_B = sequence_B\n",
    "        self.open_penalty = open_penalty\n",
    "        self.extend_penalty = extend_penalty\n",
    "        self.local = local\n",
    "        self.sub_alignments = sub_alignments\n",
    "        self.display_all_solutions = display_all_solutions\n",
    "        self.solutions = []\n",
    "        # The maximum number of characters displayed on a line\n",
    "        self.max_line_length = max_line_length\n",
    "        \n",
    "        # Execute the algorithm\n",
    "        self.fill_matrices()\n",
    "\n",
    "    @staticmethod\n",
    "    def make_move(coordinates, move_char, reverted = False):\n",
    "        \"\"\"Return the tuple coordinates, after that the specified movement has\n",
    "        been applied to it.\n",
    "        \n",
    "        Parameters:\n",
    "            coordinates: a tuple of integers\n",
    "            move_char: one of the key of Aligner.move_values\n",
    "            reverted: indicates whether the movement has to be negated before\n",
    "                being applied.\n",
    "        \n",
    "        Return value: coordinates +- Aligned.move_values[move_char]\n",
    "        \"\"\"\n",
    "            \n",
    "        if not reverted:\n",
    "            return coordinates[0] + Aligner.move_values[move_char][0], \\\n",
    "                    coordinates[1] + Aligner.move_values[move_char][1]\n",
    "        else:\n",
    "            return coordinates[0] - Aligner.move_values[move_char][0], \\\n",
    "                    coordinates[1] - Aligner.move_values[move_char][1]\n",
    "\n",
    "    @staticmethod\n",
    "    def find_maximum(matrix):\n",
    "        \"\"\"Finds s maximum in a matrix.\n",
    "        \n",
    "        Parameters:\n",
    "            matrix: the matrix\n",
    "        \n",
    "        Return value: a tuple containing the (i, j) coordinates of the maximum\n",
    "        \"\"\"\n",
    "        maximum = float(\"-inf\")\n",
    "        for i, line in enumerate(matrix):\n",
    "            for j, value in enumerate(line):\n",
    "                if value > maximum:\n",
    "                    maximum = value\n",
    "                    i_max, j_max = i, j\n",
    "        return i_max, j_max\n",
    "    \n",
    "    def fill_matrices(self):\n",
    "        \"\"\"Applies the Smith-Waterman or the Needleman-Wunsch algorithm,\n",
    "        depending on wether we do local or global alignment.\n",
    "        \"\"\"\n",
    "        m, n = len(self.sequence_A) + 1, len(self.sequence_B) + 1\n",
    "        \n",
    "        # Construct the alignment matrix first\n",
    "        self.S = Matrix(m, n, 0)\n",
    "        self.backtrace = Matrix(m, n, [\"0\"])\n",
    "        \n",
    "        if not self.local:\n",
    "            for i in range(m):\n",
    "                self.S[i, 0] = self.open_penalty + i * self.extend_penalty\n",
    "                self.backtrace[i, 0] = [\"T\"]\n",
    "                \n",
    "            for j in range(1, n):\n",
    "                self.S[0, j] = self.open_penalty + j * self.extend_penalty\n",
    "                self.backtrace[0, j] = [\"L\"]\n",
    "        \n",
    "        self.V, self.W = deepcopy(self.S), deepcopy(self.S)\n",
    "        \n",
    "        # Compute the value of all cells of the matrix\n",
    "        for i in range(1, m):\n",
    "            for j in range(1, n):\n",
    "                self.fill_cell(i, j)\n",
    "        \n",
    "        # The first cell of the solution is the maximum if we do local alignment,\n",
    "        # or the bottom-right cell if we do global alignment\n",
    "        start_cell = self.find_maximum(self.S) if self.local else (m - 1, n - 1)\n",
    "        self.find_alignment(*start_cell)\n",
    "        \n",
    "        #Print the result\n",
    "        self.print_alignment()\n",
    "        \n",
    "        # Search for remaining subalignments\n",
    "        if self.local:\n",
    "            for k in range(self.sub_alignments):\n",
    "                self.clear_path(*start_cell, self.solutions[0][\"path\"])\n",
    "                self.solutions = []\n",
    "                start_cell = self.find_maximum(self.S)\n",
    "                self.find_alignment(*start_cell)\n",
    "                self.print_alignment(True, k)\n",
    "        \n",
    "    def fill_cell(self, i, j):\n",
    "        \"\"\"Computes the value in self.S, self.W, self.v and self.backtrace for\n",
    "        the cell at position (i, j) according to the local or global alignment\n",
    "        rules.\n",
    "        \"\"\"\n",
    "        self.V[i, j] = max(self.S[i - 1, j] + self.open_penalty,\n",
    "                self.V[i - 1, j] + self.extend_penalty)\n",
    "        self.W[i, j] = max(self.S[i, j - 1] + self.open_penalty,\n",
    "                self.W[i, j - 1] + self.extend_penalty)\n",
    "        \n",
    "        # We use a dict to find the value and the origin string for the\n",
    "        # current cell at the same time\n",
    "        choices = {\n",
    "            \"T\" : self.V[i, j],\n",
    "            \"L\" : self.W[i, j],\n",
    "            # We decrease the index for accessing the sequences because the\n",
    "            # sequences have one less elements than the matrices\n",
    "            \"D\" : self.S[i - 1, j - 1] + self.score.get_score(\n",
    "                    self.sequence_A[i - 1], self.sequence_B[j - 1])\n",
    "        }\n",
    "        \n",
    "        if self.local:\n",
    "            choices[\"0\"] = 0\n",
    "            self.V[i, j] = max(self.V[i, j], 0)\n",
    "            self.W[i, j] = max(self.W[i, j], 0)\n",
    "            \n",
    "        # Find the maximum value in the dict\n",
    "        self.S[i, j] = max(choices.values())\n",
    "        \n",
    "        # Add all origins that can lead to this maximum value in the backtrace\n",
    "        self.backtrace[i, j] = \"\".join([key for key in choices \\\n",
    "                if choices[key] == self.S[i, j]])\n",
    "\n",
    "    def clear_path(self, i, j, path):\n",
    "        \"\"\"Fills with zeros cells of the solution of a local alignment in order\n",
    "        to compute the next subalignment. It also recomputes the cells impacted\n",
    "        by the zeroing.\n",
    "        \n",
    "        Parameters:\n",
    "            i, j: the coordinates of the first cell of the path\n",
    "            path: the list of moves in this solution\n",
    "        \"\"\"\n",
    "        cleared_cells = set()\n",
    "        \n",
    "        for move in path:\n",
    "            self.S[i, j] = 0\n",
    "            self.V[i, j] = 0\n",
    "            self.W[i, j] = 0\n",
    "            cleared_cells.add((i, j))\n",
    "            i, j = self.make_move((i, j), move)\n",
    "        first_cell = (i, j)\n",
    "        \n",
    "        for i in range(first_cell[0], len(self.sequence_A) + 1):\n",
    "            for j in range(first_cell[1], len(self.sequence_B) + 1):\n",
    "                for origin in self.backtrace[i, j]:\n",
    "                    if (i, j) not in cleared_cells \\\n",
    "                            and self.make_move((i, j), origin) in cleared_cells:\n",
    "                        self.fill_cell(i, j)\n",
    "                        cleared_cells.add((i, j))\n",
    "        \n",
    "        \n",
    "    def find_alignment(self, i, j, current_solution = \"\"):\n",
    "        \"\"\"Recursively iterates on the backtrace matrix to find valid paths from\n",
    "        the cell at position (i, j) to a cell with value 0 if we do local\n",
    "        alignement, or to the cell at position (0, 0) if we do global alignment.\n",
    "        \n",
    "        All paths meeting this criteria are appended to self.solutions (or just\n",
    "        the first one if not self.display_all_solutions).\n",
    "        \n",
    "        Parameters:\n",
    "            i, j: the coordinates of the currently explored cell\n",
    "            current_solution: a string reprensenting the currently explored solution\n",
    "        \"\"\"\n",
    "        \n",
    "        # Stop if we are in a zero cell, or in the top-left cell\n",
    "        if (\"0\" in self.backtrace[i, j] and self.local) \\\n",
    "                or (i == 0 and j == 0 and not self.local):\n",
    "            self.solutions.append({\"origin\" : (i, j), \"path\" : current_solution})\n",
    "        else:\n",
    "            for possibility in self.backtrace[i, j]:\n",
    "                self.find_alignment(*self.make_move((i, j), possibility),\n",
    "                        current_solution + possibility)\n",
    "                \n",
    "                # Terminate after the first possiblity if we need only one solution\n",
    "                if not self.display_all_solutions:\n",
    "                    return\n",
    "        \n",
    "    def print_alignment(self, sub_alignment = False, sub_alignment_number = 0):\n",
    "        \"\"\"Prints all the alignments found so far in self.solutions.\n",
    "        \n",
    "        Parameters:\n",
    "            sub_alignment: indicates whether we print a sub-alignment or not.\n",
    "            sub_alignment_number: the index of the sub-alignment, if applicable.\n",
    "        \"\"\"\n",
    "        gap_char = \"-\"\n",
    "        indel_char = \" \"\n",
    "        conservation_char = \":\"\n",
    "        mutation_char = \".\"\n",
    "        # The interval between two index hints dipslayed above or below the sequence\n",
    "        index_hint_interval = 10\n",
    "        max_line_length = self.max_line_length - len(\"Sequence 1: \")\n",
    "        \n",
    "        if sub_alignment:\n",
    "            print(\"Sub-alignment No\", str(sub_alignment_number + 1) + \":\")\n",
    "        else:\n",
    "            print(len(self.solutions), \"solution\" + (\"s\" if len(self.solutions) > 1 else \"\"), \"were found.\")\n",
    "        \n",
    "        for k, solution in enumerate(self.solutions):\n",
    "            sequence_A_str, sequence_B_str, mid_str= \"\", \"\", \"\"\n",
    "            index_hints_1, index_hints_2 = \"\", \"\"\n",
    "            i, j = solution[\"origin\"]\n",
    "            i -= 1\n",
    "            j -= 1\n",
    "            \n",
    "            # Iterate over the solution backward\n",
    "            for origin in reversed(solution[\"path\"]):\n",
    "                i, j = self.make_move((i, j), origin, reverted = True)\n",
    "                \n",
    "                if origin == \"T\":\n",
    "                    sequence_A_str += str(self.sequence_A[i])\n",
    "                    sequence_B_str += gap_char\n",
    "                    mid_str += indel_char\n",
    "                elif origin == \"L\":\n",
    "                    sequence_A_str += gap_char\n",
    "                    sequence_B_str += str(self.sequence_B[j])\n",
    "                    mid_str += indel_char\n",
    "                elif origin == \"D\":\n",
    "                    sequence_A_str += str(self.sequence_A[i])\n",
    "                    sequence_B_str += str(self.sequence_B[j])\n",
    "                    mid_str += conservation_char if self.sequence_A[i] \\\n",
    "                            == self.sequence_B[j] else mutation_char\n",
    "                \n",
    "                index_hints_1 += \" \"\n",
    "                index_hints_2 += \" \"\n",
    "                if i > 0 and i % index_hint_interval == 0:\n",
    "                    # Discard some characters at the end to make room for the index\n",
    "                    # hint, and add the index hint\n",
    "                    index_hints_1 = index_hints_1[:-len(str(i))] + str(i)\n",
    "                    \n",
    "                if j > 0 and j % index_hint_interval == 0:\n",
    "                    index_hints_2 = index_hints_2[:-len(str(j))] + str(j)\n",
    "            \n",
    "            # Discard the first character, so that printed sequences start indexing\n",
    "            # at 1 rather than 0 (because the index hints are shifted to the left)\n",
    "            index_hints_1 = index_hints_1[1:]\n",
    "            index_hints_2 = index_hints_2[1:]\n",
    "            \n",
    "            if len(self.solutions) > 1:\n",
    "                print(\"Solution No \" + str(k + 1) + \":\")\n",
    "            \n",
    "            for k in range(0, ((len(sequence_A_str) - 1) // max_line_length) + 1):\n",
    "                indices = slice(k * max_line_length,\n",
    "                        min((k + 1) * max_line_length,len(sequence_A_str)))\n",
    "                print(\"           \", index_hints_1[indices])\n",
    "                print(\"Sequence 1:\", sequence_A_str[indices])\n",
    "                print(\"           \", mid_str[indices])\n",
    "                print(\"Sequence 2:\", sequence_B_str[indices])\n",
    "                print(\"           \", index_hints_2[indices])\n",
    "                print()\n",
    "                \n",
    "            print()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### The constructor (`__init__`)\n",
    "This class takes all the parameters of the algorithm as constructor arguments, and executes the algorithm inside the constructor. So all we need to do in order to compute the alignment is to create an `Aligner` object with proper arguments."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Helping function (`make_move`)\n",
    "When manipulating coordinates and movement character (such as \"D\" or \"T\"), it may be convenient to mix them seamlessly. This is the intent of this method.\n",
    "\n",
    "**Note:** this method is static since it does not rely on the internal state of the object."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### The maximum function (`find_maximum`)\n",
    "This method is used to find the cell where to start the backtracking.\n",
    "\n",
    "**Note:** this method is static since it does not rely on the internal state of the object."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### The algorithm (`fill_matrices`)\n",
    "This method executes the Smith-Waterman or the Needleman-Wunsch algorithm, by using three matrices for computing the affine gape penalty in quadratic complexity, and another one for the solution backtracking.  \n",
    "The latter one (`backtrace`) contains, in each cell, a string containing `\"0\"` if we are on the last cell of an alignment, `\"T\"` if we came from the upper cell (gap in one sequence), `\"L\"` if we came from the left cell (gap in the other sequence), and `\"D\"` if we came from the upper-left cell (match or substitution).  \n",
    "A cell may be composed of mutliple distinct directions.\n",
    "\n",
    "If we do local alognment, the `\"0\"` character stops the backtracking, even if there is other possible directions for this cell. This behavior is intended, we considere that an alignment having a score of 0 somewhere should be viewed as two separate local alignments."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Compute a cell (`fill_cell`)\n",
    "Since the code that computes the value of a cell in the four matrices is used in two different methods (`fill_matrices` and `clear_path`), I factorized it in a new method."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### The backtracking (`find_alignment`)\n",
    "This method is a recursive function that finds the path from the given coordinates to the first cell having a score of 0. It is a very basic backtracking function.\n",
    "A solution consist of a string where each character has the same semantics as in the `backtrace` matrix, and the origin, i.e. the `(i, j)` coordinates of the first cell of the local alignemnt.\n",
    "\n",
    "When a full solution is found, it is appended to `self.solutions`, which is later parsed by the display method."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Sub-alignments (`clear_path`)\n",
    "In order to find sub-alignments, one have to replace cells of the main alignment by zeros, recompute impacted cells (mainly cells in the bottom of the main alignment in the matrix) and then search for a local alignment again.\n",
    "This method does the two first steps, and lets `fill_matrices` search for a subalignment.\n",
    "We consider that a cell need to be recomputed if its value depends on an already recomputed cell.\n",
    "\n",
    "For exemple, we have to recompute the value of `S[3][4]` if its value came from a match/substitution and if `S[2][3]` has been recomputed previously."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### The display (`display_alignment`)\n",
    "This method displays all the solution found in `self.solutions`. The code is quite long, because we display the two sequences with gaps at proper places, a line of dots, colons or spaces according to the correspondence between the amino acids, and two lines indicating the index of some amino acids in the sequence.  \n",
    "We wrap these lines in order to correctly display alignements longer than the screen width."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### The main function (`main`)\n",
    "This is there that the script parameters (such as the sequence files, the gap penalties, the BLOSUM) can be changed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NEAR MAIN()\n",
      "COMPUTING\n",
      "1 solution were found.\n",
      "                 460       470        480       490       500         510       \n",
      "Sequence 1: SDEALRVTPPPTSPYLNGDSPES-ANGDMDMENVTRVRLVQFQKNTDEPMGITLKMN--ELNH--CI-\n",
      "            ::.:.........:..:...::. :........:    ......:......:....:  :...  .. \n",
      "Sequence 2: SDYATYTNSNHAAPSSNASPPEGFASHSLQTSDV----VIHRKENEGFGFVIISSLNRPESGSTITVP\n",
      "            890       900       910       9999920       930       940       950 \n",
      "\n",
      "                 520       530       540       550       560         570        \n",
      "Sequence 1: --VARIMHGGMIHRQGTLHVGDEIREINGISVANQTVEQLQKMLREMRGSITFKIVPS--YR--TQSS\n",
      "              ..::..:....:...:.:::.:...::.:..:.......:.......:.:..:.:.  ..  :...\n",
      "Sequence 2: HKIGRIIDGSPADRCAKLKVGDRILAVNGQSIINMPHADIVKLIKDAGLSVTLRIIPQEELNSPTSAP\n",
      "                  960       970       980       990      1000      1010      102\n",
      "\n",
      "             580             590       600          610       620       630 \n",
      "Sequence 1: SCERDSP-----STSRQSP-ANGHSSTNNSVSDLPSTTQP---KGRQIYVRAQFEYDPAKDDLIP\n",
      "            :.:..::     ....::: :.....:.::....:...::   .:.....:.  :.. :..:..:\n",
      "Sequence 2: SSEKQSPMAQQSPLAQQSPLAQPSPATPNSPIAQPAPPQPLQLQGHENSYRS--EVK-ARQDVKP\n",
      "            0      1030      1040      1050      1060      111070       1080\n",
      "\n",
      "\n",
      "Sub alignment No 1:\n",
      "            50       360          370       380       390         400       410 \n",
      "Sequence 1: QVLDSLEEIHALTDC---SEKDLDFLHSVFQDQHLHTLLDLYDKINTKSSPQ--IRNPPSDAVQRAKE\n",
      "            :.:...:....:.::   ::..:......:... ..:.....:......:::  ...:..........\n",
      "Sequence 2: QNLSHTEVVDILKDCPIGSETSLIIHRGGFFSP-WKTPKPIMDRWENQGSPQTSLSAPAIPQNLPFPP\n",
      "               660       670       680        690       700       710       720 \n",
      "\n",
      "                  420       430       440       450       460       470       48\n",
      "Sequence 1: VLEEISCYPENNDAKELKRILTQPHFMALLQTHDVVAHEVYSDEALRVTPPPTSPYLNGDSPESANGD\n",
      "            .:.. :..:....:.. .:.   :...   ..... ....:  :.....::.::.......:.....:\n",
      "Sequence 2: ALHR-SSFPDSTEAFD-PRK---PDPY---ELYEK-SRAIY--ESRQQVPPRTSFRMDSSGPDYKELD\n",
      "                   730           740           750         760       770       7\n",
      "\n",
      "            0       490       500       510       520       530       540       \n",
      "Sequence 1: MDMENVTRVRLVQFQKNTDEPMGITLKMNELNHCIVARIMHGGMIHRQGTLHVGDEIREINGISVANQ\n",
      "            ............... ..::: :     ...   ........:...:.:.::.:::.....::.::..\n",
      "Sequence 2: VHLRRMESGFGFRIL-GGDEP-G-----QPI---LIGAVIAMGSADRDGRLHPGDELVYVDGIPVAGK\n",
      "            80       790              800          810       820       830      \n",
      "\n",
      "            550         560       570        5580          590       600        \n",
      "Sequence 1: TVEQLQKMLRE-M-RGSITFKIVPSYRTQSSSC-ERD-SP---STSRQSPANGHSSTNNSVSDLPST-\n",
      "            :.......... . .:................: :.. ::   ::...::........::....::. \n",
      "Sequence 2: THRYVIDLMHHAARNGQVNLTVRRKVLCGGEPCPENGRSPGSVSTHHSSPRSDYATYTNSNHAAPSSN\n",
      "             840       850       860       870       880       890       900    \n",
      "\n",
      "             610 \n",
      "Sequence 1: -TQPKG\n",
      "             ..:.:\n",
      "Sequence 2: ASPPEG\n",
      "                 \n",
      "\n",
      "\n",
      "Sub alignment No 2:\n",
      "            70       780       790       800       810      \n",
      "Sequence 1: IPHTTRPPKKDEENGKNYYFVSHDQMMQDISNNEYLEYGSHEDAMYGT\n",
      "            .:.::::.:..:..:..:.:......:........::.:..::..:::\n",
      "Sequence 2: VPCTTRPHKEGEVPGVDYIFITVEDFMELEKSGALLESGTYEDNYYGT\n",
      "            140       150       160       170       180    \n",
      "\n",
      "\n",
      "Sub alignment No 3:\n",
      "               520       530       540       550       560       570        580 \n",
      "Sequence 1: VARIMHGGMIHRQGTLHVGDEIREINGISVANQTVEQLQKMLREMRGSITFKIVPSYRT-QSSSCERD\n",
      "            :.:....:...:.:...:::.:.::::.:....:..........  :...........: :.......\n",
      "Sequence 2: VLRLAEDGPAIRNGRMRVGDQIIEINGESTRDMTHARAIELIKS--GGRRVRLLLKRGTGQVPEYDEP\n",
      "                1180      1190      1200      1210        1220      1230      12\n",
      "\n",
      "                  590       600  \n",
      "Sequence 1: SPSTSRQSPANGHSSTNNSVSD\n",
      "            .:..:....:.:......:..:\n",
      "Sequence 2: APWSSPAAAAPGLPEVGVSLDD\n",
      "            40      1250         \n",
      "\n",
      "\n",
      "COMPUTED\n"
     ]
    }
   ],
   "source": [
    "def main():\n",
    "    # Script parameters\n",
    "    scoring_matrix_filename = \"blosum/blosum50.iij\"\n",
    "    sequence_A_file = \"maguk-sequences.fasta\"\n",
    "    sequence_A_id = \"O14936\"\n",
    "    sequence_B_file = \"maguk-sequences.fasta\"\n",
    "    sequence_B_id = \"Q86UL8\"\n",
    "    \n",
    "    # Get the terminal width\n",
    "    # If this line raises an error, just comment it and uncomment the following line\n",
    "    #terminal_width = int(os.popen('stty size', 'r').read().split()[1])\n",
    "    terminal_width = 80\n",
    "\n",
    "    score = Score(scoring_matrix_filename)\n",
    "    sequence_A = Sequence(sequence_A_file, sequence_A_id)\n",
    "    sequence_B = Sequence(sequence_B_file, sequence_B_id)\n",
    "\n",
    "    Aligner(score, sequence_A, sequence_B,\n",
    "            open_penalty = -12,\n",
    "            extend_penalty = -2,\n",
    "            local = True,\n",
    "            display_all_solutions = False,\n",
    "            sub_alignments = 3,\n",
    "            max_line_length = terminal_width)\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    print(\"Computing alignments...\")\n",
    "    main()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Tests\n",
    "Tests have been done during the whole coding process, but here is a sample to show that results are similar to the results given by the LALIGN server.\n",
    "\n",
    "**File used**: `maguk-sequences.fasta`  \n",
    "**Sequence A:** `O14936`  \n",
    "**Sequence B:** `Q86UL8`  \n",
    "**Gap penalty:** opening: -12, extend: -2  \n",
    "**Substitution matrix:** BLOSUM 50  \n",
    "** Results:**   \n",
    "```\n",
    "1 solution were found.\n",
    "                 460       470       480       490       500         510        \n",
    "Sequence 1: SDEALRVTPPPTSPYLNGDSPESANGDMDMENVTRVRLVQFQKNTDEPMGITLKMN--ELNH--CI--\n",
    "            ::.:.........:..:...::.. ..... ....: ......:......:....:  :...  ..  \n",
    "Sequence 2: SDYATYTNSNHAAPSSNASPPEGF-ASHSL-QTSDV-VIHRKENEGFGFVIISSLNRPESGSTITVPH\n",
    "            890       900       9910        9920       930       940       950  \n",
    "\n",
    "                520       530       540       550       560         570         \n",
    "Sequence 1: -VARIMHGGMIHRQGTLHVGDEIREINGISVANQTVEQLQKMLREMRGSITFKIVPS--YR--TQSSS\n",
    "             ..::..:....:...:.:::.:...::.:..:.......:.......:.:..:.:.  ..  :...:\n",
    "Sequence 2: KIGRIIDGSPADRCAKLKVGDRILAVNGQSIINMPHADIVKLIKDAGLSVTLRIIPQEELNSPTSAPS\n",
    "                 960       970       980       990      1000      1010      1020\n",
    "\n",
    "            580             590       600          610       620       630 \n",
    "Sequence 1: CERDSP-----STSRQSP-ANGHSSTNNSVSDLPSTTQP---KGRQIYVRAQFEYDPAKDDLIP\n",
    "            .:..::     ....::: :.....:.::....:...::   .:.....:.  :.. :..:..:\n",
    "Sequence 2: SEKQSPMAQQSPLAQQSPLAQPSPATPNSPIAQPAPPQPLQLQGHENSYRS--EVK-ARQDVKP\n",
    "                  1030      1040      1050      1060      111070       1080\n",
    "\n",
    "\n",
    "Sub alignment No 1:\n",
    "            50       360          370       380       390         400       410 \n",
    "Sequence 1: QVLDSLEEIHALTDC---SEKDLDFLHSVFQDQHLHTLLDLYDKINTKSSPQ--IRNPPSDAVQRAKE\n",
    "            :.:...:....:.::   ::..: ..:.........:.....:......:::  ...:..........\n",
    "Sequence 2: QNLSHTEVVDILKDCPIGSETSL-IIHRGGFFSPWKTPKPIMDRWENQGSPQTSLSAPAIPQNLPFPP\n",
    "               660       670        680       690       700       710       720 \n",
    "\n",
    "                  420       430       440       450       460       470       48\n",
    "Sequence 1: VLEEISCYPENNDAKELKRILTQPHFMALLQTHDVVAHEVYSDEALRVTPPPTSPYLNGDSPESANGD\n",
    "            .:.. :..:....:.. .:.   :...   ..... ....:  :.....::.::.......:.....:\n",
    "Sequence 2: ALHR-SSFPDSTEAFD-PRK---PDPY---ELYEK-SRAIY--ESRQQVPPRTSFRMDSSGPDYKELD\n",
    "                   730           740           750         760       770       7\n",
    "\n",
    "            0       490       500       510       520       530       540       \n",
    "Sequence 1: MDMENVTRVRLVQFQKNTDEPMGITLKMNELNHCIVARIMHGGMIHRQGTLHVGDEIREINGISVANQ\n",
    "            ............... ..:::     . ...   ........:...:.:.::.:::.....::.::..\n",
    "Sequence 2: VHLRRMESGFGFRIL-GGDEP-----G-QPI---LIGAVIAMGSADRDGRLHPGDELVYVDGIPVAGK\n",
    "            80       790              800          810       820       830      \n",
    "\n",
    "            550         560       570        5580          590       600        \n",
    "Sequence 1: TVEQLQKMLRE-M-RGSITFKIVPSYRTQSSSC-ERD-SP---STSRQSPANGHSSTNNSVSDLPST-\n",
    "            :.......... . .:................: :.. ::   ::...::........::....::. \n",
    "Sequence 2: THRYVIDLMHHAARNGQVNLTVRRKVLCGGEPCPENGRSPGSVSTHHSSPRSDYATYTNSNHAAPSSN\n",
    "             840       850       860       870       880       890       900    \n",
    "\n",
    "             610 \n",
    "Sequence 1: -TQPKG\n",
    "             ..:.:\n",
    "Sequence 2: ASPPEG\n",
    "                 \n",
    "\n",
    "\n",
    "Sub alignment No 2:\n",
    "            70       780       790       800       810      \n",
    "Sequence 1: IPHTTRPPKKDEENGKNYYFVSHDQMMQDISNNEYLEYGSHEDAMYGT\n",
    "            .:.::::.:..:..:..:.:......:........::.:..::..:::\n",
    "Sequence 2: VPCTTRPHKEGEVPGVDYIFITVEDFMELEKSGALLESGTYEDNYYGT\n",
    "            140       150       160       170       180    \n",
    "\n",
    "\n",
    "Sub alignment No 3:\n",
    "               520       530       540       550       560       570        580 \n",
    "Sequence 1: VARIMHGGMIHRQGTLHVGDEIREINGISVANQTVEQLQKMLREMRGSITFKIVPSYRT-QSSSCERD\n",
    "            :.:....:...:.:...:::.:.::::.:....:.........  .:...........: :.......\n",
    "Sequence 2: VLRLAEDGPAIRNGRMRVGDQIIEINGESTRDMTHARAIELIK--SGGRRVRLLLKRGTGQVPEYDEP\n",
    "                1180      1190      1200      1210        1220      1230      12\n",
    "\n",
    "                  590       600  \n",
    "Sequence 1: SPSTSRQSPANGHSSTNNSVSD\n",
    "            .:..:....:.:......:..:\n",
    "Sequence 2: APWSSPAAAAPGLPEVGVSLDD\n",
    "            40      1250         \n",
    "```\n",
    "**LALIGN results:**  \n",
    "```\n",
    ">>B 1455 bp                                               (1455 aa)\n",
    " Waterman-Eggert score: 155;  44.9 bits; E(1) <  4e-08\n",
    "28.9% identity (57.8% similar) in 128 aa overlap (515-632:954-1081)\n",
    "\n",
    "          520       530       540       550       560       570    \n",
    "A      VARIMHGGMIHRQGTLHVGDEIREINGISVANQTVEQLQKMLREMRGSITFKIVPSYR--\n",
    "       ..::. :.   : . :.:::.:  .:: :. :.   .. :....   :.:..:.:. .  \n",
    "B      IGRIIDGSPADRCAKLKVGDRILAVNGQSIINMPHADIVKLIKDAGLSVTLRIIPQEELN\n",
    "           960       970       980       990      1000      1010   \n",
    "\n",
    "              580            590        600       610       620    \n",
    "A      --TQSSSCERDSPST-----SRQSP-ANGHSSTNNSVSDLPSTTQPKGRQIYVRAQFEYD\n",
    "         :.. : :..:: .     ..::: :.   .: ::    :.  ::   : .  .     \n",
    "B      SPTSAPSSEKQSPMAQQSPLAQQSPLAQPSPATPNSPIAQPAPPQPLQLQGHENSYRSEV\n",
    "          1020      1030      1040      1050      1060      1070   \n",
    "\n",
    "          630  \n",
    "A      PAKDDLIP\n",
    "        :..:. :\n",
    "B      KARQDVKP\n",
    "          1080 \n",
    "\n",
    ">--\n",
    " Waterman-Eggert score: 124;  36.9 bits; E(1) <  1e-05\n",
    "36.5% identity (67.3% similar) in 52 aa overlap (767-818:134-185)\n",
    "\n",
    "        770       780       790       800       810        \n",
    "A      FAYPIPHTTRPPKKDEENGKNYYFVSHDQMMQDISNNEYLEYGSHEDAMYGT\n",
    "       .   .: :::: :. :  : .: :.. ...:.  ...  :: :..:: .:::\n",
    "B      YLRTVPCTTRPHKEGEVPGVDYIFITVEDFMELEKSGALLESGTYEDNYYGT\n",
    "           140       150       160       170       180     \n",
    "\n",
    ">--\n",
    " Waterman-Eggert score: 119;  35.6 bits; E(1) <  2.5e-05\n",
    "22.5% identity (52.1% similar) in 280 aa overlap (349-612:655-910)\n",
    "\n",
    "      350       360          370       380       390        400    \n",
    "A      QVLDSLEEIHALTDC---SEKDLDFLHSVFQDQHLHTLLDLYDKINTKSSPQIR-NPPSD\n",
    "       : :.  : .  : ::   :: .: ..:     .  .:   ..:. ....:::   . :. \n",
    "B      QNLSHTEVVDILKDCPIGSETSL-IIHRGGFFSPWKTPKPIMDRWENQGSPQTSLSAPAI\n",
    "          660       670        680       690       700       710   \n",
    "\n",
    "          410       420       430       440       450       460    \n",
    "A      AVQRAKEVLEEISCYPENNDAKELKRILTQPHFMALLQTHDVVAHEVYSDEALRVTPPPT\n",
    "         .       . : .:....: . ..    :  . : .     .. .:  :. . .:: :\n",
    "B      PQNLPFPPALHRSSFPDSTEAFDPRK----PDPYELYEK----SRAIY--ESRQQVPPRT\n",
    "           720       730           740           750         760   \n",
    "\n",
    "          470       480          490       500       510       520 \n",
    "A      SPYLNGDSPESANGDMD---MENVTRVRLVQFQKNTDEPMGITLKMNELNHCIVARIMHG\n",
    "       :  .....:.  . :.    ::.    :..    . ::: :  .        ... ..  \n",
    "B      SFRMDSSGPDYKELDVHLRRMESGFGFRIL----GGDEP-GQPI--------LIGAVIAM\n",
    "           770       780       790            800               810\n",
    "\n",
    "             530       540       550         560       570         \n",
    "A      GMIHRQGTLHVGDEIREINGISVANQTVEQLQKMLREM--RGSITFKIVPSYRTQSSSCE\n",
    "       :   :.: :: :::.  ..:: ::..: . .  ....    :.... .  .    .  : \n",
    "B      GSADRDGRLHPGDELVYVDGIPVAGKTHRYVIDLMHHAARNGQVNLTVRRKVLCGGEPCP\n",
    "              820       830       840       850       860       870\n",
    "\n",
    "     580            590       600         610  \n",
    "A      RD--SP---STSRQSPANGHSSTNNSVSDLPST--TQPKG\n",
    "       ..  ::   :: ..:: . ... .::    ::.  . :.:\n",
    "B      ENGRSPGSVSTHHSSPRSDYATYTNSNHAAPSSNASPPEG\n",
    "              880       890       900       910\n",
    "```\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can see that the first line of the first alignment does not appear in the LALIGN results. It may be due to a different criterion for stopping a local alignment, which is thus more strict in LALIGN than in our implementation.\n",
    "\n",
    "Also, the first and the second sub-alignments are swapped in the two results. This is certainly due to a different zeroing of the matrix after the main alignment, or to an optimization in LALIGN that has knock-on effects on the second and third maximums in the alignment matrix."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
