{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# INFO-F208: project 3, PSSM profiles\n",
    "The goal of this project is to make a position-specific scoring matrix (*PSSM*) from aligned sequences of amino acids of the WW domain. The sequences that will be used come from human proteins only.\n",
    "\n",
    "A PSSM is a compact representation of multiple aligned sequences, and is commonly used to represent domains of sequences, such as SH-3, PDZ, or WW in our case. PSSMs can be used to align an unknown sequence to a whole domain, with only one pass of the alignment algorithm, in order to check whether the sequence belongs to the domain.\n",
    "\n",
    "## Data\n",
    "As explained in the statement, I got sequences of the WW domain from the SMART database in a `.fasta` file, and I aligned them with two online aligner. I chose MUSCLE and CLUSTAL Omega. The aligned domains has been saved into `msaresults-MUSCLE.fasta` and `msaresults-CLUSTAL.fasta`.\n",
    "\n",
    "In order to calculate the PSSM, one need $p(a)$, i.e. the probability of the amino acid $a$ in the random pattern. This information has been gathered from the Swiss-prot database statistics page, and is encoded in the global dictionnary `p` in following the Python code.\n",
    "\n",
    "## Choices in the implementation\n",
    "A few decisions has to be taken in this project regarding the details of the implementation.\n",
    "### Pseudo-counts\n",
    "The goal of pseudo-counts is to avoid errors when using logarithms in the computing of the PSSM: in the case where the frequency ($f_{u,a}$) is equal to zero in the formula\n",
    "$$m_{u,a} = \\log\\frac{f_{u,a}}{p(a)}$$\n",
    "we have an error ($\\log(0)\\notin \\mathbb{R}$).  \n",
    "So the solution is to use pseudo-counts, and change the formula to\n",
    "$$m_{u,a} = \\log\\frac{\\alpha f_{u,a} + \\beta p(a)}{(\\alpha + \\beta)p(a)}$$\n",
    "So that with positive $\\alpha$ and $\\beta$, the fraction is never null.  \n",
    "I chose $\\alpha$ as the number of sequences without gap at the position $u$. Consequently, $\\alpha$ is a function of $u$, thus we need to calculate a value of $\\alpha$ for each column.  \n",
    "I also chose $\\beta = \\sqrt{N_{seq}}$.\n",
    "### The logarithm base\n",
    "The base of the logarithm is not explicitly given in the slides, so I assumed a base 2. Nonetheless, this can easily be changed by setting the global variable `logBase` in the Python code.\n",
    "\n",
    "### Various notes on the code\n",
    "* The code below outputs the PSSM profile of the MUSCLE alignment, but you can get the CLUSTAL PSSM as well by setting the following variable to `\"clustal\"`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "usedAligner = \"muscle\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* If the code errors and says that the maximum recursion depth has been reached, uncomment the following lines:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#import sys\n",
    "#sys.setrecursionlimit(1500)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Here you can set the number of character displayed per line on the alignment output."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "terminal_width = 80"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* The code outputs first the PSSM, then the alignment of both sequences in test.fasta with the PSSM, and finally the consensus sequence of the PSSM\n",
    "* The PSSM output is wide, so it is not very convenient in jupyter. The solution is to copy-paste to a text editor, so that the PSSM can be easily inspected.\n",
    "* Since this project needs an adaptated version of the first project, I copied and then adapted the code of the first project into this one. I put \"`### FILE project1.py ###`\" and \"`### END FILE project1.py ###`\" at the bounds of the first project code.\n",
    "\n",
    "## Code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  |     0 |     1 |     2 |     3 |     4 |     5 |     6 |     7 |     8 |     9 |    10 |    11 |    12 |    13 |    14 |    15 |    16 |    17 |    18 |    19 |    20 |    21 |    22 |    23 |    24 |    25 |    26 |    27 |    28 |    29 |    30 |    31 |    32 |    33 |    34 |    35 |    36 |    37 |    38 |    39 |    40 |    41 |    42 |    43 |    44 |    45 |    46 |    47 |    48 |    49 |    50 |    51 |    52 |    53 |    54 |    55 |    56 |    57 |    58 |    59 | \n",
      "  +-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------\n",
      "A | -1.56 | -1.38 | -2.33 | -0.02 | -3.23 | -4.06 | -4.06 | -1.57 | 0.701 | -4.06 | -0.94 | -0.08 | 0.163 | -1.38 | -1.54 | -4.06 | -1.07 | -3.23 | -4.06 | -4.06 | -1.38 | -4.06 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.32 | -2.03 | -1.22 | -4.06 | -1.38 | -3.23 | -1.07 | -4.05 | -1.78 | -4.06 | -4.05 | -1.22 | -1.78 | \n",
      "R | -4.03 | -2.34 | -1.57 | 0.127 | -1.90 | -4.06 | -1.30 | 0.716 | 2.265 | -1.08 | -4.06 | -0.08 | 0.215 | -0.29 | -1.54 | -0.55 | 2.957 | 0.590 | -4.06 | -4.06 | -4.06 | -4.06 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.26 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.70 | -0.55 | -1.90 | 2.440 | 0.887 | -1.57 | -0.17 | -4.05 | -1.90 | 1.890 | -4.05 | 2.494 | -0.71 | \n",
      "N | -1.54 | -2.70 | -0.00 | -2.01 | 0.354 | -4.06 | -4.06 | -4.06 | -1.55 | -2.01 | -4.06 | -0.08 | 0.244 | 1.740 | -1.54 | -2.01 | -4.06 | -4.06 | -4.06 | -4.06 | -4.06 | 3.657 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.22 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.35 | -1.19 | 2.900 | 2.097 | 0.125 | -4.06 | -4.05 | -4.05 | -4.05 | -2.70 | -0.00 | -4.05 | -4.05 | -1.55 | \n",
      "D | 1.104 | -4.06 | -2.32 | -0.27 | -1.29 | -4.06 | -4.06 | -4.06 | -2.94 | -2.32 | 3.013 | -0.08 | 0.050 | 1.768 | -1.54 | -1.89 | -1.56 | -2.94 | -4.06 | -4.06 | -4.06 | 2.647 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.36 | -0.39 | -0.39 | -0.39 | -0.39 | -0.26 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -2.30 | 0.540 | -0.54 | -1.06 | -1.89 | -4.05 | -4.05 | -4.05 | 0.734 | 2.458 | -4.05 | 0.050 | 1.008 | \n",
      "C | -0.69 | -4.06 | -4.06 | 0.788 | -4.06 | -4.06 | 0.535 | -0.70 | 0.788 | -4.06 | -0.70 | -0.08 | -4.05 | -4.06 | 0.591 | -4.06 | -1.56 | 0.535 | -4.06 | -4.06 | -4.06 | -4.06 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | 0.062 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.16 | -4.06 | -4.06 | -4.06 | -0.16 | -0.69 | 0.534 | 1.003 | -4.06 | -4.06 | -4.05 | 0.534 | 0.787 | \n",
      "E | 0.500 | -3.10 | -0.81 | 0.856 | 0.179 | -4.06 | 3.111 | 2.176 | -1.32 | -0.23 | -2.53 | -0.08 | 0.611 | 0.317 | -1.24 | -1.80 | -0.43 | -4.06 | -4.06 | -4.06 | -4.06 | -4.06 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.29 | -0.39 | -0.39 | -0.39 | -0.39 | -0.29 | -0.39 | -0.39 | -0.39 | -0.39 | -0.96 | 0.557 | -0.05 | -0.96 | 0.900 | -4.05 | -1.80 | -4.05 | 2.527 | -3.10 | -1.80 | -1.32 | 0.442 | \n",
      "Q | -0.87 | -4.06 | 0.040 | -1.97 | -0.64 | -4.06 | -0.43 | 1.193 | -4.06 | 0.770 | -4.06 | -0.08 | 0.040 | 1.131 | -1.54 | -4.06 | 0.686 | -1.16 | -2.67 | -4.06 | -4.06 | -4.06 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.07 | -0.08 | -0.39 | -0.39 | -0.35 | -0.22 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.35 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -3.98 | 0.770 | -1.97 | 1.193 | 0.289 | -4.05 | 2.872 | -4.05 | 0.997 | -4.06 | -4.05 | 0.289 | -4.05 | \n",
      "G | 0.432 | -2.57 | -1.60 | -0.29 | 3.003 | -4.06 | -1.86 | -4.06 | 0.313 | -4.06 | -4.06 | -0.08 | -1.60 | -0.39 | -1.54 | 3.455 | -0.20 | -3.13 | -4.06 | -4.06 | -4.06 | -4.06 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.37 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.27 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.37 | -0.39 | -0.39 | -3.98 | -1.86 | -1.86 | 1.108 | -4.06 | -3.13 | -3.13 | -4.05 | -4.06 | -4.06 | -4.05 | -1.38 | -0.29 | \n",
      "H | -4.03 | -4.06 | -1.33 | -4.06 | -1.33 | -4.06 | -4.06 | -4.06 | 2.088 | 0.649 | 1.051 | -0.08 | -0.82 | 1.051 | -1.11 | -4.06 | -4.06 | -0.82 | -4.06 | -4.06 | 1.267 | 0.302 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.10 | -0.39 | -0.39 | -0.39 | -0.31 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | 4.510 | -4.06 | -4.06 | -2.13 | -4.06 | -4.05 | -2.13 | -4.05 | 1.163 | 2.607 | -4.05 | 0.092 | -4.05 | \n",
      "I | -0.64 | -1.65 | -4.06 | -4.06 | -1.98 | -4.06 | -2.40 | -0.97 | 0.202 | 1.283 | -4.06 | -0.08 | -0.64 | -2.40 | -1.54 | -1.98 | -1.39 | 0.789 | -4.06 | -4.06 | 2.103 | -0.80 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -3.98 | 0.989 | -1.98 | -4.06 | -0.38 | -4.05 | -4.05 | -1.65 | -0.26 | -4.06 | -4.05 | -1.65 | -1.65 | \n",
      "L | -0.80 | 2.839 | -4.06 | -1.95 | -3.33 | -4.06 | -2.20 | -4.06 | -1.75 | -1.57 | -3.33 | -0.08 | -1.57 | -1.13 | -1.54 | -4.06 | -1.75 | -4.06 | -3.33 | -4.06 | -0.90 | -4.06 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.32 | -0.37 | -0.39 | -0.39 | -0.37 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -1.74 | -1.57 | -2.84 | -2.48 | -1.01 | -4.05 | -2.84 | -4.05 | 0.445 | -0.90 | -4.05 | -3.32 | 1.181 | \n",
      "K | -1.96 | -4.06 | -0.13 | -1.14 | -1.96 | -4.06 | -0.13 | 1.347 | 0.057 | 1.015 | -4.06 | -0.08 | -0.62 | -1.14 | -1.54 | -0.62 | 1.015 | -0.62 | -4.06 | -4.06 | -4.06 | -4.06 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -1.95 | -1.37 | -4.06 | 2.213 | 0.519 | -2.99 | -4.05 | -4.05 | 0.306 | 0.919 | -4.05 | 0.645 | -0.77 | \n",
      "M | -4.03 | -4.06 | -2.19 | -4.06 | -0.53 | -4.06 | 0.010 | 2.309 | -4.06 | -0.23 | -4.06 | -0.08 | -2.19 | -0.90 | -1.54 | -4.06 | -1.41 | -2.19 | -1.41 | -4.06 | -4.06 | -4.06 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -3.98 | -4.06 | -2.19 | -4.06 | -4.06 | -4.05 | 0.404 | -4.05 | -4.06 | -1.41 | -4.05 | -2.19 | 1.079 | \n",
      "F | 0.423 | -4.06 | -4.06 | -4.06 | -4.06 | -4.06 | -4.06 | -4.06 | -4.06 | 1.156 | -4.06 | -0.08 | -0.23 | -4.06 | -0.86 | -2.65 | -1.95 | -1.49 | 2.530 | 3.043 | 0.064 | -4.06 | -0.07 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.35 | -0.39 | -1.13 | -0.61 | -4.06 | -4.06 | -4.06 | -4.05 | -4.05 | 0.950 | -4.06 | -2.65 | -4.05 | -0.85 | -1.95 | \n",
      "P | 2.646 | -0.07 | 3.857 | 2.822 | 0.670 | -4.06 | -4.06 | -1.38 | -4.06 | -4.06 | -4.06 | -0.08 | 2.292 | -4.06 | -1.54 | -4.06 | -4.06 | 1.785 | -4.06 | -4.06 | -4.06 | -4.06 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.35 | -0.35 | -0.35 | -0.35 | -0.39 | -0.24 | -0.39 | 0.594 | 0.148 | -4.06 | -4.06 | -2.83 | -4.05 | -4.05 | -4.05 | -4.06 | 0.249 | 4.290 | -2.82 | -0.20 | \n",
      "S | -0.02 | -0.20 | -1.30 | -0.11 | -0.30 | -4.06 | -0.78 | -1.78 | -2.51 | -1.78 | 1.014 | -0.08 | 0.529 | 1.207 | -0.67 | -2.51 | -4.06 | -1.10 | -4.06 | -4.06 | -4.06 | -1.78 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.39 | -0.29 | -0.29 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.37 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.29 | -0.39 | -0.39 | -0.29 | -1.29 | -2.51 | 0.792 | -0.94 | -0.02 | 1.585 | 0.471 | -4.05 | -2.10 | -0.52 | -4.05 | -0.11 | -0.78 | \n",
      "T | -0.02 | -0.02 | -4.06 | 0.344 | -4.06 | -4.06 | 0.344 | -0.37 | -1.87 | 1.086 | 2.333 | -0.08 | -0.51 | -0.13 | -0.16 | -4.06 | -0.02 | 1.466 | -4.06 | -4.06 | -2.93 | -4.06 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.26 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | 0.171 | 0.260 | 3.162 | -4.06 | 2.692 | 3.568 | 2.842 | -4.05 | 0.172 | -4.06 | -4.05 | 0.700 | 1.085 | \n",
      "W | -4.03 | -4.06 | -4.06 | 0.147 | -4.06 | 6.407 | -0.39 | -4.06 | -4.06 | 2.405 | -4.06 | -0.08 | -4.05 | -4.06 | -1.54 | -4.06 | 1.107 | -4.06 | -4.06 | 2.201 | 1.324 | -4.06 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.39 | -0.39 | -0.39 | -0.23 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -3.98 | -4.06 | -4.06 | -4.06 | -4.06 | -4.05 | -4.05 | 6.126 | -4.06 | -4.06 | -4.05 | -4.05 | 1.828 | \n",
      "Y | -1.63 | -4.06 | -4.06 | 0.699 | -4.06 | -0.78 | -0.24 | 0.302 | 1.186 | 1.612 | -4.06 | -0.08 | -1.14 | -4.06 | -1.54 | 0.447 | -0.78 | 1.101 | 4.579 | 4.303 | 2.879 | -4.06 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.77 | -4.06 | -4.06 | -4.06 | -4.06 | -4.05 | -4.05 | 1.101 | -4.06 | -4.06 | -4.05 | -1.14 | -0.24 | \n",
      "V | -2.54 | -0.46 | -2.14 | -3.11 | -1.15 | -4.06 | -0.16 | -0.57 | 0.290 | 0.531 | -3.11 | -0.08 | -1.15 | -1.35 | -1.54 | -3.11 | -0.70 | 1.957 | -2.55 | -4.06 | 2.056 | -1.56 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.39 | -0.37 | -0.39 | -0.39 | -0.29 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.29 | -0.39 | -0.39 | -3.08 | -0.70 | -4.06 | -4.06 | -0.25 | 0.223 | -1.34 | -4.05 | -4.06 | -3.11 | -4.05 | 0.916 | -0.46 | \n",
      "\n",
      "\n",
      "********************************************************************************\n",
      "Alignment of D6C652\n",
      "1 solution were found.\n",
      "                 150         160                                170  \n",
      "Sequence : PLPPGWEMAKT-PS-GQRYFLN-------------------------HIDQTTTWQDPR\n",
      "           ::::::::... :. :..:..:                         :...::.:..::\n",
      "Consensus: PLPPGWEMRWDRPDCGRVYYYNFTLVAQAHSSQVLCGENDTRESVPSHNTRTTQWEHPR\n",
      "                   10        20        30        40        50        \n",
      "\n",
      "\n",
      "Sub alignment No 1:\n",
      "           00        210                                  220        230 \n",
      "Sequence : PLPDGWEQALT-PE-GE-AYFIN-------------------------HKN-KSTSWLDPRL\n",
      "           :::.:::.... :. :.  :..:                         : : ..:.:..::.\n",
      "Consensus: PLPPGWEMRWDRPDCGRV-YYYNFTLVAQAHSSQVLCGENDTRESVPSH-NTRTTQWEHPRW\n",
      "                   10         20        30        40         50         \n",
      "\n",
      "\n",
      "Sub alignment No 2:\n",
      "               390          \n",
      "Sequence : ITQSTIPTQ-QNRFPDYL\n",
      "           .:. :  :: . . :. .\n",
      "Consensus: NTR-T--TQWE-H-PR-W\n",
      "           50               \n",
      "\n",
      "\n",
      "Sub alignment No 3:\n",
      "           20        330       340         350       360       370       380   \n",
      "Sequence : PVCTTG-ISQELRTMTMNSSDPFLNSG-TY-HSRDESTESGLSMSSYSVPRTPDDFLNSVDEMDTGEA\n",
      "           : .  :   .. :. ..    :  ..: .: .  . .:......::.......:........ .:.. \n",
      "Consensus: P-P--GW--EM-RW-DR----P--DCGRVYYY--N-FTLVAQAHSSQVLCGENDTRESVPSH-NTRT-\n",
      "                        10              20           30        40         50   \n",
      "\n",
      "                 390\n",
      "Sequence : ITQ-ST-IPT\n",
      "            ::  .  :.\n",
      "Consensus: -TQW-EH-PR\n",
      "                    \n",
      "\n",
      "\n",
      "********************************************************************************\n",
      "Alignment of P46935\n",
      "1 solution were found.\n",
      "           60       470                                  480       490 \n",
      "Sequence : PLPPGWEERTHT-D-GRVFFIN-------------------------HNIKKTQWEDPRL\n",
      "           :::::::.:... : :::...:                         ::...::::.::.\n",
      "Consensus: PLPPGWEMRWDRPDCGRVYYYNFTLVAQAHSSQVLCGENDTRESVPSHNTRTTQWEHPRW\n",
      "                   10        20        30        40        50         \n",
      "\n",
      "\n",
      "Sub alignment No 1:\n",
      "           50       260                                  270         \n",
      "Sequence : PLPPGWEERQDV-L-GRTYYVN-------------------------HESRRTQWKRP\n",
      "           :::::::.:.:. . ::.::.:                         :..:.:::..:\n",
      "Consensus: PLPPGWEMRWDRPDCGRVYYYNFTLVAQAHSSQVLCGENDTRESVPSHNTRTTQWEHP\n",
      "                   10        20        30        40        50       \n",
      "\n",
      "\n",
      "Sub alignment No 2:\n",
      "             410         420                                430        \n",
      "Sequence : GLPPGWEEKQD--DRGRSYYVD-------------------------HNSKTTTW-SKPTM\n",
      "           .::::::...:  :.::.::..                         ::..::.:  .:..\n",
      "Consensus: PLPPGWEMRWDRPDCGRVYYYNFTLVAQAHSSQVLCGENDTRESVPSHNTRTTQWE-HPRW\n",
      "                   10        20        30        40        50          \n",
      "\n",
      "\n",
      "Sub alignment No 3:\n",
      "            130        140        150        160           170       180       \n",
      "Sequence : LNPKWNEEILFRV-LPQ-RHRILFEVF-DENRLTRDDF---L-GQVDVPLYPLPTENPRMERPYT-FK\n",
      "           :.:.: :.   :. .:. . :. .  . ..........   : :..:.......... : .   : . \n",
      "Consensus: LPPGW-EM---RWDRPDCG-RV-Y--YYNFTLVAQAHSSQVLCGENDTRESVPSHNT-R-T---TQW-\n",
      "                      10            20        30        40         50          \n",
      "\n",
      "            190  \n",
      "Sequence : DFVLHPR\n",
      "           .   :::\n",
      "Consensus: E---HPR\n",
      "                 \n",
      "\n",
      "\n",
      "PSSM consensus sequence: PLPPGWEMRWDRPDCGRVYYYNFTLVAQAHSSQVLCGENDTRESVPSHNTRTTQWEHPRW\n"
     ]
    }
   ],
   "source": [
    "from math import log\n",
    "from copy import deepcopy\n",
    "import os, sys\n",
    "from enum import Enum\n",
    "\n",
    "### FILE project1.py ###\n",
    "\n",
    "AminoAcid = Enum(\"AminoAcid\", \"A R N D C E Q G H I L K M F P S T W Y V B Z X\")\n",
    "# Make the str() function print only the amino acid letter by redefining __str__\n",
    "AminoAcid.__str__ = lambda self: self.name[-1]\n",
    "\n",
    "class Sequence:\n",
    "    \"\"\"Represents a sequence of amino acids.\n",
    "    It is suited to compute the best alignment between two sequences.\"\"\"\n",
    "    \n",
    "    def __init__(self, filename, sequence_id):\n",
    "        \"\"\"Constructor. Loads the sequence from a file.\n",
    "        \n",
    "        Parameters:\n",
    "            filename: the filename of the file containing the sequence.\n",
    "            sequence_id: the identifier of the sequence in the file.\n",
    "                This is the string between the two first bars \"|\" on the line\n",
    "                preceding the sequence.\n",
    "            \"\"\"\n",
    "        self.load_from_file(filename, sequence_id)\n",
    "    \n",
    "    def __getitem__(self, index):\n",
    "        \"\"\"Returns the elements at the specified index. Allows the syntax\n",
    "        sequence[i].\n",
    "        \"\"\"\n",
    "        return self.sequence[index]\n",
    "        \n",
    "    def __len__(self):\n",
    "        \"\"\"Returns the lenght of the sequence. Allows the syntax len(sequence).\"\"\"\n",
    "        return len(self.sequence)\n",
    "\n",
    "    def load_from_file(self, filename, sequence_id):\n",
    "        \"\"\"Loads the sequence data from a file.\n",
    "        \n",
    "        Parameters:\n",
    "            filename: the filename of the file containing the sequence.\n",
    "            sequence_id: the identifier of the sequence in the file.\n",
    "                This is the string between the two first bars \"|\" on the line\n",
    "                preceding the sequence.\n",
    "        \"\"\"\n",
    "        found_sequence = False\n",
    "        sequence = \"\"\n",
    "        with open(filename) as file:\n",
    "            for line in file:\n",
    "                if not found_sequence:\n",
    "                    # We found the sequence if this is the right identifier line\n",
    "                    found_sequence = (line[0] == \">\"\n",
    "                            and line.split(\"|\")[1] == sequence_id)\n",
    "                else:\n",
    "                    if line[0] != \">\":\n",
    "                        sequence += line\n",
    "                    else:\n",
    "                        # Stop adding lines to the sequence\n",
    "                        found_sequence = False\n",
    "                        \n",
    "        if sequence == \"\":\n",
    "            raise RuntimeError(\"Sequence not found in the specified file.\")\n",
    "\n",
    "        try:\n",
    "            sequence = [AminoAcid[character] for character in sequence \\\n",
    "                    if not character.isspace()]\n",
    "        except KeyError as e:\n",
    "            raise RuntimeError(\"An amino acid is not valid in the sequence: \"\n",
    "                    + str(e))\n",
    "\n",
    "        # We didn't worked on self.sequence directly to preserve the internal\n",
    "        # state in case of failure\n",
    "        self.sequence = sequence\n",
    "        \n",
    "class Matrix:\n",
    "    \"\"\"Represents a m*n matrix.\n",
    "    It is used to abstract the implementation of matrices, and allow possible\n",
    "    optimisations easily, without having to change the whole code.\n",
    "    All lines don't have to have the same length, so it may be used to represent\n",
    "    a triangular matrix.\n",
    "    \n",
    "    \"\"\"\n",
    "    def __init__(self, m = 0, n = 0, value=None):\n",
    "        \"\"\"Construct the matrix\n",
    "        \n",
    "        Parameters:\n",
    "            m, n: height and width of the matrix\n",
    "            value: the initial value to put in the cells.\n",
    "        \"\"\"\n",
    "        self.matrix = [[value] * n for i in range(m)]\n",
    "    \n",
    "    def __getitem__(self, key):\n",
    "        if isinstance(key, tuple):\n",
    "            return self.matrix[key[0]][key[1]]\n",
    "        else:\n",
    "            return self.matrix[key]\n",
    "            \n",
    "    def __setitem__(self, key, value):\n",
    "        if isinstance(key, tuple):\n",
    "            self.matrix[key[0]][key[1]] = value\n",
    "        else:\n",
    "            self.matrix[key] = value\n",
    "    \n",
    "    def __str__(self):\n",
    "        cell_width = max([max([len(str(c)) for c in line]) for line in self.matrix])\n",
    "        res = \"\"\n",
    "        for line in self.matrix:\n",
    "            for cell in line:\n",
    "                res += str(cell).rjust(cell_width) + \" \"\n",
    "            res += \"\\n\"\n",
    "        return res\n",
    "            \n",
    "    def append(self, iterable):\n",
    "        \"\"\"Adds a line to the bottom of the matrix.\"\"\"\n",
    "        self.matrix.append(list(iterable))\n",
    "    \n",
    "class Aligner:\n",
    "    \"\"\"Holds functions and variable related to the sequence alignment.\"\"\"\n",
    "    \n",
    "    move_values = {\"D\" : (-1, -1), \"T\" : (-1, 0), \"L\" : (0, -1), \"0\" : (0, 0)}\n",
    "\n",
    "    def __init__(self, sequence, profile,\n",
    "            gap_penalties,\n",
    "            sub_alignments = 0,\n",
    "            display_all_solutions = True,\n",
    "            max_line_length = 80):\n",
    "        \"\"\"Constructor, with all the parameters of the algorithm.\"\"\"\n",
    "        self.sequence = sequence\n",
    "        self.profile = profile\n",
    "        self.gap_penalties = gap_penalties\n",
    "        self.sub_alignments = sub_alignments\n",
    "        self.display_all_solutions = display_all_solutions\n",
    "        self.solutions = []\n",
    "        # The maximum number of characters displayed on a line\n",
    "        self.max_line_length = max_line_length\n",
    "        \n",
    "        # Execute the algorithm\n",
    "        self.fill_matrices()\n",
    "        \n",
    "\n",
    "    @staticmethod\n",
    "    def make_move(coordinates, move_char, reverted = False):\n",
    "        \"\"\"Return the tuple coordinates, after that the specified movement has\n",
    "        been applied to it.\n",
    "        \n",
    "        Parameters:\n",
    "            coordinates: a tuple of integers\n",
    "            move_char: one of the key of Aligner.move_values\n",
    "            reverted: indicates whether the movement has to be negated before\n",
    "                being applied.\n",
    "        \n",
    "        Return value: coordinates +- Aligned.move_values[move_char]\n",
    "        \"\"\"\n",
    "            \n",
    "        if not reverted:\n",
    "            return coordinates[0] + Aligner.move_values[move_char][0], \\\n",
    "                    coordinates[1] + Aligner.move_values[move_char][1]\n",
    "        else:\n",
    "            return coordinates[0] - Aligner.move_values[move_char][0], \\\n",
    "                    coordinates[1] - Aligner.move_values[move_char][1]\n",
    "\n",
    "    @staticmethod\n",
    "    def find_maximum(matrix):\n",
    "        \"\"\"Finds s maximum in a matrix.\n",
    "        \n",
    "        Parameters:\n",
    "            matrix: the matrix\n",
    "        \n",
    "        Return value: a tuple containing the (i, j) coordinates of the maximum\n",
    "        \"\"\"\n",
    "        maximum = float(\"-inf\")\n",
    "        for i, line in enumerate(matrix):\n",
    "            for j, value in enumerate(line):\n",
    "                if value > maximum:\n",
    "                    maximum = value\n",
    "                    i_max, j_max = i, j\n",
    "        return i_max, j_max\n",
    "    \n",
    "    def fill_matrices(self):\n",
    "        \"\"\"Fill the matrices in order to compute the alignment.\"\"\"\n",
    "        m, n = len(self.sequence) + 1, len(self.profile) + 1\n",
    "        \n",
    "        # Construct the alignment matrix first\n",
    "        self.S = Matrix(m, n, 0)\n",
    "        self.backtrace = Matrix(m, n, [\"0\"])\n",
    "        \n",
    "        # Compute the value of all cells of the matrix\n",
    "        for i in range(1, m):\n",
    "            for j in range(1, n):\n",
    "                self.fill_cell(i, j)\n",
    "        \n",
    "        # The first cell of the solution is the maximum cell\n",
    "        start_cell = self.find_maximum(self.S)\n",
    "        self.find_alignment(*start_cell)\n",
    "        \n",
    "        #Print the result\n",
    "        self.print_alignment()\n",
    "        \n",
    "        # Search for remaining subalignments\n",
    "        for k in range(self.sub_alignments):\n",
    "            self.clear_path(*start_cell, self.solutions[0][\"path\"])\n",
    "            self.solutions = []\n",
    "            start_cell = self.find_maximum(self.S)\n",
    "            self.find_alignment(*start_cell)\n",
    "            self.print_alignment(True, k)\n",
    "        \n",
    "    def fill_cell(self, i, j):\n",
    "        \"\"\"Computes the value in self.S and self.backtrace for the cell at\n",
    "        position (i, j) according to the alignment rules.\n",
    "        \"\"\"\n",
    "        \n",
    "        # We use a dict to find the value and the origin string for the\n",
    "        # current cell at the same time\n",
    "        choices = {\n",
    "            \"T\" : self.S[i - 1, j] + self.gap_penalties[j - 1],\n",
    "            \"L\" : self.S[i, j - 1] + self.gap_penalties[j - 1],\n",
    "            # We decrease the index for accessing the sequences because the\n",
    "            # sequences have one less elements than the matrices\n",
    "            \"D\" : self.S[i - 1, j - 1] + self.profile[j - 1, self.sequence[i - 1]],\n",
    "            \"0\" : 0\n",
    "        }\n",
    "            \n",
    "        # Find the maximum value in the dict\n",
    "        self.S[i, j] = max(choices.values())\n",
    "        \n",
    "        # Add all origins that can lead to this maximum value in the backtrace\n",
    "        self.backtrace[i, j] = \"\"\n",
    "        for key in choices:\n",
    "            if choices[key] == self.S[i, j]:\n",
    "                self.backtrace[i,j] += key\n",
    "\n",
    "    def clear_path(self, i, j, path):\n",
    "        \"\"\"Fills with zeros cells of the solution of a local alignment in order\n",
    "        to compute the next subalignment. It also recomputes the cells impacted\n",
    "        by the zeroing.\n",
    "        \n",
    "        Parameters:\n",
    "            i, j: the coordinates of the first cell of the path\n",
    "            path: the list of moves in this solution\n",
    "        \"\"\"\n",
    "        cleared_cells = set()\n",
    "        \n",
    "        for move in path:\n",
    "            self.S[i, j] = 0\n",
    "            cleared_cells.add((i, j))\n",
    "            i, j = self.make_move((i, j), move)\n",
    "        first_cell = (i, j)\n",
    "        \n",
    "        for i in range(first_cell[0], len(self.sequence) + 1):\n",
    "            for j in range(first_cell[1], len(self.profile) + 1):\n",
    "                for origin in self.backtrace[i, j]:\n",
    "                    if (i, j) not in cleared_cells \\\n",
    "                            and self.make_move((i, j), origin) in cleared_cells:\n",
    "                        self.fill_cell(i, j)\n",
    "                        cleared_cells.add((i, j))\n",
    "        \n",
    "        \n",
    "    def find_alignment(self, i, j, current_solution = \"\"):\n",
    "        \"\"\"Recursively iterates on the backtrace matrix to find valid paths from\n",
    "        the cell at position (i, j) to a cell with value 0.\n",
    "        \n",
    "        All paths meeting this criteria are appended to self.solutions (or just\n",
    "        the first one if not self.display_all_solutions).\n",
    "        \n",
    "        Parameters:\n",
    "            i, j: the coordinates of the currently explored cell\n",
    "            current_solution: a string reprensenting the currently explored solution\n",
    "        \"\"\"\n",
    "        # Stop if we are in a zero cell\n",
    "        if \"0\" in self.backtrace[i, j]:\n",
    "            self.solutions.append({\"origin\" : (i, j), \"path\" : current_solution})\n",
    "        else:\n",
    "            for possibility in self.backtrace[i, j]:\n",
    "                self.find_alignment(*self.make_move((i, j), possibility),\n",
    "                        current_solution + possibility)\n",
    "                \n",
    "                # Terminate after the first possiblity if we need only one solution\n",
    "                if not self.display_all_solutions:\n",
    "                    return\n",
    "        \n",
    "    def print_alignment(self, sub_alignment = False, sub_alignment_number = 0):\n",
    "        \"\"\"Prints all the alignments found so far in self.solutions.\n",
    "        \n",
    "        Parameters:\n",
    "            sub_alignment: indicates whether we print a sub-alignment or not.\n",
    "            sub_alignment_number: the index of the sub-alignment, if applicable.\n",
    "        \"\"\"\n",
    "        gap_char = \"-\"\n",
    "        indel_char = \" \"\n",
    "        conservation_char = \":\"\n",
    "        mutation_char = \".\"\n",
    "        # The interval between two index hints dipslayed above or below the sequence\n",
    "        index_hint_interval = 10\n",
    "        max_line_length = self.max_line_length - len(\"Sequence 1: \")\n",
    "        \n",
    "        if sub_alignment:\n",
    "            print(\"Sub alignment No\", str(sub_alignment_number + 1) + \":\")\n",
    "        else:\n",
    "            print(len(self.solutions), \"solution\" + (\"s\" if len(self.solutions) > 1 else \"\"), \"were found.\")\n",
    "            \n",
    "        for k, solution in enumerate(self.solutions):\n",
    "            sequence_str, profile_str, mid_str= \"\", \"\", \"\"\n",
    "            index_hints_1, index_hints_2 = \"\", \"\"\n",
    "            i, j = solution[\"origin\"]\n",
    "            i -= 1\n",
    "            j -= 1\n",
    "            \n",
    "            # Iterate over the solution backward\n",
    "            for origin in reversed(solution[\"path\"]):\n",
    "                i, j = self.make_move((i, j), origin, reverted = True)\n",
    "                \n",
    "                if origin == \"T\":\n",
    "                    sequence_str += str(self.sequence[i])\n",
    "                    profile_str += gap_char\n",
    "                    mid_str += indel_char\n",
    "                elif origin == \"L\":\n",
    "                    sequence_str += gap_char\n",
    "                    profile_str += str(self.profile.consensus[j])\n",
    "                    mid_str += indel_char\n",
    "                elif origin == \"D\":\n",
    "                    sequence_str += str(self.sequence[i])\n",
    "                    profile_str += str(self.profile.consensus[j])\n",
    "                    mid_str += conservation_char if str(self.sequence[i]) \\\n",
    "                            == self.profile.consensus[j] else mutation_char\n",
    "                \n",
    "                index_hints_1 += \" \"\n",
    "                index_hints_2 += \" \"\n",
    "                if i > 0 and i % index_hint_interval == 0 and origin != \"L\":\n",
    "                    # Discard some characters at the end to make room for the index\n",
    "                    # hint, and add the index hint\n",
    "                    index_hints_1 = index_hints_1[:-len(str(i))] + str(i)\n",
    "                    \n",
    "                if j > 0 and j % index_hint_interval == 0 and origin != \"T\":\n",
    "                    index_hints_2 = index_hints_2[:-len(str(j))] + str(j)\n",
    "            \n",
    "            # Discard the first character, so that printed sequences start indexing\n",
    "            # at 1 rather than 0 (because the index hints are shifted to the left)\n",
    "            index_hints_1 = index_hints_1[1:]\n",
    "            index_hints_2 = index_hints_2[1:]\n",
    "            \n",
    "            if len(self.solutions) > 1:\n",
    "                print(\"Solution No \" + str(k + 1) + \":\")\n",
    "            \n",
    "            for k in range(0, ((len(sequence_str) - 1) // max_line_length) + 1):\n",
    "                indices = slice(k * max_line_length,\n",
    "                        min((k + 1) * max_line_length,len(sequence_str)))\n",
    "                print(\"          \", index_hints_1[indices])\n",
    "                print(\"Sequence :\", sequence_str[indices])\n",
    "                print(\"          \", mid_str[indices])\n",
    "                print(\"Consensus:\", profile_str[indices])\n",
    "                print(\"          \", index_hints_2[indices])\n",
    "                print()\n",
    "                \n",
    "            print()\n",
    "\n",
    "### END FILE project1.py ###\n",
    "\n",
    "gap = \"-\"\n",
    "aminoAcids = \"ARNDCEQGHILKMFPSTWYV\"\n",
    "\n",
    "# The following probabilities has been gathered from the Swiss-prot database\n",
    "# statistics page\n",
    "p={ \"E\" : 0.0674, \"H\" : 0.0227, \"S\" : 0.066,  \"K\" : 0.0582, \"Y\" : 0.0292,\n",
    "\t\"T\" : 0.0535, \"W\" : 0.0109, \"L\" : 0.0965, \"R\" : 0.0553, \"M\" : 0.0241,\n",
    "\t\"Q\" : 0.0393, \"A\" : 0.0826, \"V\" : 0.0687, \"F\" : 0.0386, \"C\" : 0.0137,\n",
    "\t\"N\" : 0.0406, \"P\" : 0.0472, \"I\" : 0.0593, \"D\" : 0.0546, \"G\" : 0.0708}\n",
    "\n",
    "logBase = 2\n",
    "\n",
    "class Pssm:\n",
    "    def __init__(self, filename):\n",
    "        self.load(filename)\n",
    "    \n",
    "    def load(self, filename):\n",
    "        sequences = []\n",
    "        \n",
    "        with open(filename) as file:\n",
    "            for line in file:\n",
    "                line = line.strip()\n",
    "                if not line.startswith(\">\"):\n",
    "                    sequences.append(line.upper())\n",
    "        \n",
    "        self.numberColumns = len(sequences[0])\n",
    "        matrix = [{aa : 0 for aa in aminoAcids} for _ in range(self.numberColumns)]\n",
    "        beta = len(sequences) ** 0.5\n",
    "        self.matrix = deepcopy(matrix)\n",
    "        \n",
    "        for sequence in sequences:\n",
    "            for i, aa in enumerate(sequence):\n",
    "                if aa in aminoAcids:\n",
    "                    matrix[i][aa] += 1\n",
    "                    \n",
    "        \n",
    "        for i, column in enumerate(matrix):\n",
    "            alpha = sum(column.values())\n",
    "            for aa, count in column.items():\n",
    "                frequency = count / len(sequences)\n",
    "                probability = (alpha * frequency + beta * p[aa]) / (alpha + beta)\n",
    "                self.matrix[i][aa] = log(probability / p[aa], logBase)\n",
    "        \n",
    "        self.computeConsensus()\n",
    "    \n",
    "    def __getitem__(self, t):\n",
    "        return self.matrix[int(t[0])][str(t[1])]\n",
    "    \n",
    "    def __len__(self):\n",
    "        return len(self.matrix)\n",
    "    \n",
    "    def __str__(self):\n",
    "        res = \"  | \"\n",
    "        cellWidth = 5\n",
    "        for i in range(self.numberColumns):\n",
    "            res += str(i).rjust(cellWidth) + \" | \"\n",
    "        res += \"\\n  \" + \"+-------\" * self.numberColumns + \"\\n\"\n",
    "        for aa in aminoAcids:\n",
    "            res += aa + \" | \"\n",
    "            for column in self.matrix:\n",
    "                res += str(column[aa])[:cellWidth].rjust(cellWidth) + \" | \"\n",
    "            res += \"\\n\"\n",
    "        res += \"\\n\"\n",
    "        return res\n",
    "    \n",
    "    def computeConsensus(self):\n",
    "        self.consensus = \"\"\n",
    "        for column in self.matrix:\n",
    "            self.consensus += max(column, key=lambda aa : column[aa])\n",
    "\n",
    "def main():\n",
    "    # Script parameters\n",
    "    if usedAligner == \"muscle\":\n",
    "        msaFilename = \"msaresults-MUSCLE.fasta\"\n",
    "    else:\n",
    "        msaFilename = \"msaresults-CLUSTAL.fasta\"\n",
    "        \n",
    "    sequence_file = \"test.fasta\"\n",
    "    sequence_ids = [\"D6C652\", \"P46935\"]\n",
    "    \n",
    "    pssm = Pssm(msaFilename)\n",
    "    \n",
    "    print(pssm)\n",
    "    \n",
    "    for sequence_id in sequence_ids:\n",
    "        sequence = Sequence(sequence_file, sequence_id)\n",
    "        print(\"*\" * terminal_width)\n",
    "        print(\"Alignment of\", sequence_id)\n",
    "        Aligner(sequence, pssm, gap_penalties = [-1] * len(pssm),\n",
    "                display_all_solutions = False, sub_alignments = 3,\n",
    "                max_line_length = terminal_width)\n",
    "    \n",
    "    print(\"PSSM consensus sequence:\", pssm.consensus)\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Results analysis\n",
    "\n",
    "### PSSM construction\n",
    "Both PSSMs has been computed and can be found in `MUSCLE-PSSM.txt` and `CLUSTAL-PSSM.txt`. They aren't shown here because of their excessive width. We will compare the PSSM files and the weblogo constructed from the alignment fasta files. We will also give the consensus sequence of each PSSM profile.\n",
    "\n",
    "Of course these comparisons are not formal and do not prove that this project works for all domain alignments. The aim is just to convince that it works for this case, and it is sufficient, because this project targets the WW domain only.\n",
    "\n",
    "#### MUSCLE Weblogo\n",
    "\n",
    "![MUSCLE-logo](MUSCLE-logo.png)\n",
    "The consensus sequence of the corresponding PSSM:\n",
    "\n",
    "```\n",
    "PLPPGWEMRWDRPDCGRVYYYNFTLVAQAHSSQVLCGENDTRESVPSHNTRTTQWEHPRW\n",
    "```\n",
    "\n",
    "The consensus does not always agree with the weblogo for positions with low score, but for position with high score the consensus matches the weblogo.  \n",
    "\n",
    "By searching for high values on the weblogo and in `MUSCLE-PSSM.txt`, we can see that they all match. For exemple:\n",
    "\n",
    "* A big W on the weblogo at index 5, and a value of 6.407 for W in the PSSM at this index\n",
    "* D slightly bigger than T at index 10, and 3.013 for D and 2.333 for T in the PSSM  at this index\n",
    "* A big P on the weblogo at index 57, and a value of 4.290 for P in the PSSM at this index\n",
    "\n",
    "#### CLUSTAL Weblogo\n",
    "\n",
    "![CLUSTAL-logo](CLUSTAL-logo.png)\n",
    "The consensus sequence of the corresponding PSSM:\n",
    "\n",
    "```\n",
    "PLPPGWEMRWDPSNGRVYYYNHNTQKVCGENWTRLGSLQESVPSYNHINRTTQWEHPRW\n",
    "```\n",
    "\n",
    "The consensus does not always agree with the weblogo for positions with low score, but for position with high score the consensus matches the weblogo.\n",
    "\n",
    "By searching for high values on the weblogo and in `CLUSTAL-PSSM.txt`, we can see that they all match. For exemple:\n",
    "* A big W on the weblogo at index 5, and a value of 6.407 for W in the PSSM at this index\n",
    "* A big G on the weblogo at index 14 and a value of 3.455 for G in the PSSM at this index\n",
    "* A big W on the weblogo at index 53 and a value of 6.127 for W in the PSSM at this index\n",
    "\n",
    "#### Preserved positions in the Weblogos\n",
    "\n",
    "There are a few amino acids that are very well conserved in all sequences of the domain:\n",
    "\n",
    "* Amino acid P at index 2\n",
    "* Amino acid W at index 5\n",
    "* Amino acid Y at index 17/18\n",
    "* Amino acid Y at index 18/19\n",
    "* Amino acid H at index 21/47\n",
    "* Amino acid W at index 53/54\n",
    "* Amino acid P at index 56/57\n",
    "\n",
    "Note that the most prominent amino acids can be found in both alignment, but their exact indices slightly differ. This is due to different gaps in the alignments. The most noticeable difference is for an amino acid H that is before a big gap in the CLUSTAL alignment, and right after that gap in the MUSCLE alignment.\n",
    "#### HMM logo\n",
    "Here is the HMM logo for the WW family:  \n",
    "\n",
    "![HMM-logo](HMM-logo.png)\n",
    "\n",
    "We can see that the preserved positions are the same as in the weblogos, although scaled differently. Gaps are shown in the HMM logo with vertical red lines, and by the last two lines of numbers below the image. The gap information in the HMM logo roughly matches with the observed gaps in the Weblogos. Because the weblogos match our PSSMs, we can conclude that our PSSMs also match the HMM logo.\n",
    "\n",
    "### PSSM-sequence alignment\n",
    "From the output of the code we can see that the alignment algorithm detects 2 WW domains in the protein YAP1A_XENLA (*D6C652*), and three WW domains in the protein NEDD4_MOUSE (*P46935*). The remaining sub-alignments are irrelevant, they just show that there is no other WW domains alignments.\n",
    "\n",
    "For the YAP1A_XENLA, the output shows that the first WW domain starts at index 142, and the second at index 199. The same information can be found in the UNIPROT database:\n",
    "\n",
    "![UNIPROT-YAP1A_XENLA](UNIPROT-YAP1A_XENLA.png)\n",
    "\n",
    "For the NEDD4_MOUSE, the output shows that the first WW domain starts at index 459, the second at index 249 and the third at index 406. The same information can be found in the UNIPROT database:\n",
    "![UNIPROT-NEDD4_MOUSE](UNIPROT-NEDD4_MOUSE.png)\n",
    "\n",
    "The index in the UNIPROT database has sometimes a difference of 1 with the index we got, this is due to different start index (the output of this code shows indices starting from 1, not from 0).\n",
    "\n",
    "## Conclusion\n",
    "All the results of this project are very similar to the ones given by online tools and databases. Thus we can conclude that this project successfully met the requirements of the statement."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2+"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
