{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# INFO-F208: project 3, PSSM profiles\n",
    "The goal of this project is to make a position-specific scoring matrix (*PSSM*) from aligned sequences of amino acids of the WW domain. The sequences that will be used come from human proteins only.\n",
    "\n",
    "A PSSM is a compact representation of multiple aligned sequences, and is commonly used to represent domains of sequences, such as SH-3, PDZ, or WW in our case. PSSMs can be used to align an unknown sequence to a whole domain, with only one pass of the alignment algorithm, in order to check whether the sequence belongs to the domain.\n",
    "\n",
    "## Data\n",
    "As explained in the statement, I got sequences of the WW domain from the SMART database in a `.fasta` file, and I aligned them with two online aligner. I chose MUSCLE and CLUSTAL Omega. The aligned domains has been saved into `msaresults-MUSCLE.fasta` and `msaresults-CLUSTAL.fasta`.\n",
    "\n",
    "In order to calculate the PSSM, one need $p(a)$, i.e. the probability of the amino acid $a$ in the random pattern. This information has been gathered from the Swiss-prot database statistics page, and is encoded in the global dictionnary `p` in following the Python code.\n",
    "\n",
    "## Choices in the implementation\n",
    "A few decisions has to be taken in this project regarding the details of the implementation.\n",
    "### Pseudo-counts\n",
    "The goal of pseudo-counts is to avoid errors when using logarithms in the computing of the PSSM: in the case where the frequency ($f_{u,a}$) is equal to zero in the formula\n",
    "$$m_{u,a} = \\log\\frac{f_{u,a}}{p(a)}$$\n",
    "we have an error ($\\log(0)\\notin \\mathbb{R}$).  \n",
    "So the solution is to use pseudo-counts, and change the formula to\n",
    "$$m_{u,a} = \\log\\frac{\\alpha f_{u,a} + \\beta p(a)}{(\\alpha + \\beta)p(a)}$$\n",
    "So that with positive $\\alpha$ and $\\beta$, the fraction is never null.  \n",
    "I chose $\\alpha$ as the number of sequences without gap at the position $u$. Consequently, $\\alpha$ is a function of $u$, thus we need to calculate a value of $\\alpha$ for each column.  \n",
    "I also chose $\\beta = \\sqrt{N_{seq}}$.\n",
    "### The logarithm base\n",
    "The base of the logarithm is not explicitly given in the slides, so I assumed a base 2. Nonetheless, this can easily be changed by setting the global variable `logBase` in the beginning of the Python code.\n",
    "\n",
    "### Output of the code\n",
    "The code below outputs the PSSM profile of the MUSCLE alignment, but one can get the CLUSTAL PSSM as well by commenting and uncommenting the appropriate lines in the `main` fonction.  \n",
    "The output is wide, so it is not very convenient in jupyter. The solution is to copy-paste the output to a text editor, so that the PSSM can be easily inspected.\n",
    "\n",
    "## Code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  |     0 |     1 |     2 |     3 |     4 |     5 |     6 |     7 |     8 |     9 |    10 |    11 |    12 |    13 |    14 |    15 |    16 |    17 |    18 |    19 |    20 |    21 |    22 |    23 |    24 |    25 |    26 |    27 |    28 |    29 |    30 |    31 |    32 |    33 |    34 |    35 |    36 |    37 |    38 |    39 |    40 |    41 |    42 |    43 |    44 |    45 |    46 |    47 |    48 |    49 |    50 |    51 |    52 |    53 |    54 |    55 |    56 |    57 |    58 |    59 | \n",
      "  +-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------\n",
      "A | -1.56 | -1.38 | -2.33 | -0.02 | -3.23 | -4.06 | -4.06 | -1.57 | 0.701 | -4.06 | -0.94 | -0.08 | 0.163 | -1.38 | -1.54 | -4.06 | -1.07 | -3.23 | -4.06 | -4.06 | -1.38 | -4.06 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.32 | -2.03 | -1.22 | -4.06 | -1.38 | -3.23 | -1.07 | -4.05 | -1.78 | -4.06 | -4.05 | -1.22 | -1.78 | \n",
      "R | -4.03 | -2.34 | -1.57 | 0.127 | -1.90 | -4.06 | -1.30 | 0.716 | 2.265 | -1.08 | -4.06 | -0.08 | 0.215 | -0.29 | -1.54 | -0.55 | 2.957 | 0.590 | -4.06 | -4.06 | -4.06 | -4.06 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.26 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.70 | -0.55 | -1.90 | 2.440 | 0.887 | -1.57 | -0.17 | -4.05 | -1.90 | 1.890 | -4.05 | 2.494 | -0.71 | \n",
      "N | -1.54 | -2.70 | -0.00 | -2.01 | 0.354 | -4.06 | -4.06 | -4.06 | -1.55 | -2.01 | -4.06 | -0.08 | 0.244 | 1.740 | -1.54 | -2.01 | -4.06 | -4.06 | -4.06 | -4.06 | -4.06 | 3.657 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.22 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.35 | -1.19 | 2.900 | 2.097 | 0.125 | -4.06 | -4.05 | -4.05 | -4.05 | -2.70 | -0.00 | -4.05 | -4.05 | -1.55 | \n",
      "D | 1.104 | -4.06 | -2.32 | -0.27 | -1.29 | -4.06 | -4.06 | -4.06 | -2.94 | -2.32 | 3.013 | -0.08 | 0.050 | 1.768 | -1.54 | -1.89 | -1.56 | -2.94 | -4.06 | -4.06 | -4.06 | 2.647 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.36 | -0.39 | -0.39 | -0.39 | -0.39 | -0.26 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -2.30 | 0.540 | -0.54 | -1.06 | -1.89 | -4.05 | -4.05 | -4.05 | 0.734 | 2.458 | -4.05 | 0.050 | 1.008 | \n",
      "C | -0.69 | -4.06 | -4.06 | 0.788 | -4.06 | -4.06 | 0.535 | -0.70 | 0.788 | -4.06 | -0.70 | -0.08 | -4.05 | -4.06 | 0.591 | -4.06 | -1.56 | 0.535 | -4.06 | -4.06 | -4.06 | -4.06 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | 0.062 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.16 | -4.06 | -4.06 | -4.06 | -0.16 | -0.69 | 0.534 | 1.003 | -4.06 | -4.06 | -4.05 | 0.534 | 0.787 | \n",
      "E | 0.500 | -3.10 | -0.81 | 0.856 | 0.179 | -4.06 | 3.111 | 2.176 | -1.32 | -0.23 | -2.53 | -0.08 | 0.611 | 0.317 | -1.24 | -1.80 | -0.43 | -4.06 | -4.06 | -4.06 | -4.06 | -4.06 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.29 | -0.39 | -0.39 | -0.39 | -0.39 | -0.29 | -0.39 | -0.39 | -0.39 | -0.39 | -0.96 | 0.557 | -0.05 | -0.96 | 0.900 | -4.05 | -1.80 | -4.05 | 2.527 | -3.10 | -1.80 | -1.32 | 0.442 | \n",
      "Q | -0.87 | -4.06 | 0.040 | -1.97 | -0.64 | -4.06 | -0.43 | 1.193 | -4.06 | 0.770 | -4.06 | -0.08 | 0.040 | 1.131 | -1.54 | -4.06 | 0.686 | -1.16 | -2.67 | -4.06 | -4.06 | -4.06 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.07 | -0.08 | -0.39 | -0.39 | -0.35 | -0.22 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.35 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -3.98 | 0.770 | -1.97 | 1.193 | 0.289 | -4.05 | 2.872 | -4.05 | 0.997 | -4.06 | -4.05 | 0.289 | -4.05 | \n",
      "G | 0.432 | -2.57 | -1.60 | -0.29 | 3.003 | -4.06 | -1.86 | -4.06 | 0.313 | -4.06 | -4.06 | -0.08 | -1.60 | -0.39 | -1.54 | 3.455 | -0.20 | -3.13 | -4.06 | -4.06 | -4.06 | -4.06 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.37 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.27 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.37 | -0.39 | -0.39 | -3.98 | -1.86 | -1.86 | 1.108 | -4.06 | -3.13 | -3.13 | -4.05 | -4.06 | -4.06 | -4.05 | -1.38 | -0.29 | \n",
      "H | -4.03 | -4.06 | -1.33 | -4.06 | -1.33 | -4.06 | -4.06 | -4.06 | 2.088 | 0.649 | 1.051 | -0.08 | -0.82 | 1.051 | -1.11 | -4.06 | -4.06 | -0.82 | -4.06 | -4.06 | 1.267 | 0.302 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.10 | -0.39 | -0.39 | -0.39 | -0.31 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | 4.510 | -4.06 | -4.06 | -2.13 | -4.06 | -4.05 | -2.13 | -4.05 | 1.163 | 2.607 | -4.05 | 0.092 | -4.05 | \n",
      "I | -0.64 | -1.65 | -4.06 | -4.06 | -1.98 | -4.06 | -2.40 | -0.97 | 0.202 | 1.283 | -4.06 | -0.08 | -0.64 | -2.40 | -1.54 | -1.98 | -1.39 | 0.789 | -4.06 | -4.06 | 2.103 | -0.80 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -3.98 | 0.989 | -1.98 | -4.06 | -0.38 | -4.05 | -4.05 | -1.65 | -0.26 | -4.06 | -4.05 | -1.65 | -1.65 | \n",
      "L | -0.80 | 2.839 | -4.06 | -1.95 | -3.33 | -4.06 | -2.20 | -4.06 | -1.75 | -1.57 | -3.33 | -0.08 | -1.57 | -1.13 | -1.54 | -4.06 | -1.75 | -4.06 | -3.33 | -4.06 | -0.90 | -4.06 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.32 | -0.37 | -0.39 | -0.39 | -0.37 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -1.74 | -1.57 | -2.84 | -2.48 | -1.01 | -4.05 | -2.84 | -4.05 | 0.445 | -0.90 | -4.05 | -3.32 | 1.181 | \n",
      "K | -1.96 | -4.06 | -0.13 | -1.14 | -1.96 | -4.06 | -0.13 | 1.347 | 0.057 | 1.015 | -4.06 | -0.08 | -0.62 | -1.14 | -1.54 | -0.62 | 1.015 | -0.62 | -4.06 | -4.06 | -4.06 | -4.06 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -1.95 | -1.37 | -4.06 | 2.213 | 0.519 | -2.99 | -4.05 | -4.05 | 0.306 | 0.919 | -4.05 | 0.645 | -0.77 | \n",
      "M | -4.03 | -4.06 | -2.19 | -4.06 | -0.53 | -4.06 | 0.010 | 2.309 | -4.06 | -0.23 | -4.06 | -0.08 | -2.19 | -0.90 | -1.54 | -4.06 | -1.41 | -2.19 | -1.41 | -4.06 | -4.06 | -4.06 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -3.98 | -4.06 | -2.19 | -4.06 | -4.06 | -4.05 | 0.404 | -4.05 | -4.06 | -1.41 | -4.05 | -2.19 | 1.079 | \n",
      "F | 0.423 | -4.06 | -4.06 | -4.06 | -4.06 | -4.06 | -4.06 | -4.06 | -4.06 | 1.156 | -4.06 | -0.08 | -0.23 | -4.06 | -0.86 | -2.65 | -1.95 | -1.49 | 2.530 | 3.043 | 0.064 | -4.06 | -0.07 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.35 | -0.39 | -1.13 | -0.61 | -4.06 | -4.06 | -4.06 | -4.05 | -4.05 | 0.950 | -4.06 | -2.65 | -4.05 | -0.85 | -1.95 | \n",
      "P | 2.646 | -0.07 | 3.857 | 2.822 | 0.670 | -4.06 | -4.06 | -1.38 | -4.06 | -4.06 | -4.06 | -0.08 | 2.292 | -4.06 | -1.54 | -4.06 | -4.06 | 1.785 | -4.06 | -4.06 | -4.06 | -4.06 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.35 | -0.35 | -0.35 | -0.35 | -0.39 | -0.24 | -0.39 | 0.594 | 0.148 | -4.06 | -4.06 | -2.83 | -4.05 | -4.05 | -4.05 | -4.06 | 0.249 | 4.290 | -2.82 | -0.20 | \n",
      "S | -0.02 | -0.20 | -1.30 | -0.11 | -0.30 | -4.06 | -0.78 | -1.78 | -2.51 | -1.78 | 1.014 | -0.08 | 0.529 | 1.207 | -0.67 | -2.51 | -4.06 | -1.10 | -4.06 | -4.06 | -4.06 | -1.78 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.39 | -0.29 | -0.29 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.37 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.29 | -0.39 | -0.39 | -0.29 | -1.29 | -2.51 | 0.792 | -0.94 | -0.02 | 1.585 | 0.471 | -4.05 | -2.10 | -0.52 | -4.05 | -0.11 | -0.78 | \n",
      "T | -0.02 | -0.02 | -4.06 | 0.344 | -4.06 | -4.06 | 0.344 | -0.37 | -1.87 | 1.086 | 2.333 | -0.08 | -0.51 | -0.13 | -0.16 | -4.06 | -0.02 | 1.466 | -4.06 | -4.06 | -2.93 | -4.06 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.26 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | 0.171 | 0.260 | 3.162 | -4.06 | 2.692 | 3.568 | 2.842 | -4.05 | 0.172 | -4.06 | -4.05 | 0.700 | 1.085 | \n",
      "W | -4.03 | -4.06 | -4.06 | 0.147 | -4.06 | 6.407 | -0.39 | -4.06 | -4.06 | 2.405 | -4.06 | -0.08 | -4.05 | -4.06 | -1.54 | -4.06 | 1.107 | -4.06 | -4.06 | 2.201 | 1.324 | -4.06 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.39 | -0.39 | -0.39 | -0.23 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -3.98 | -4.06 | -4.06 | -4.06 | -4.06 | -4.05 | -4.05 | 6.126 | -4.06 | -4.06 | -4.05 | -4.05 | 1.828 | \n",
      "Y | -1.63 | -4.06 | -4.06 | 0.699 | -4.06 | -0.78 | -0.24 | 0.302 | 1.186 | 1.612 | -4.06 | -0.08 | -1.14 | -4.06 | -1.54 | 0.447 | -0.78 | 1.101 | 4.579 | 4.303 | 2.879 | -4.06 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.77 | -4.06 | -4.06 | -4.06 | -4.06 | -4.05 | -4.05 | 1.101 | -4.06 | -4.06 | -4.05 | -1.14 | -0.24 | \n",
      "V | -2.54 | -0.46 | -2.14 | -3.11 | -1.15 | -4.06 | -0.16 | -0.57 | 0.290 | 0.531 | -3.11 | -0.08 | -1.15 | -1.35 | -1.54 | -3.11 | -0.70 | 1.957 | -2.55 | -4.06 | 2.056 | -1.56 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.08 | -0.39 | -0.37 | -0.39 | -0.39 | -0.29 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.39 | -0.29 | -0.39 | -0.39 | -3.08 | -0.70 | -4.06 | -4.06 | -0.25 | 0.223 | -1.34 | -4.05 | -4.06 | -3.11 | -4.05 | 0.916 | -0.46 | \n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from math import log\n",
    "from copy import deepcopy\n",
    "\n",
    "gap = \"-\"\n",
    "aminoAcids = \"ARNDCEQGHILKMFPSTWYV\"\n",
    "\n",
    "# The following probabilities has been gathered from the Swiss-prot database\n",
    "# statistics page\n",
    "p={ \"E\" : 0.0674, \"H\" : 0.0227, \"S\" : 0.066,  \"K\" : 0.0582, \"Y\" : 0.0292,\n",
    "    \"T\" : 0.0535, \"W\" : 0.0109, \"L\" : 0.0965, \"R\" : 0.0553, \"M\" : 0.0241,\n",
    "    \"Q\" : 0.0393, \"A\" : 0.0826, \"V\" : 0.0687, \"F\" : 0.0386, \"C\" : 0.0137,\n",
    "    \"N\" : 0.0406, \"P\" : 0.0472, \"I\" : 0.0593, \"D\" : 0.0546, \"G\" : 0.0708}\n",
    "\n",
    "logBase = 2\n",
    "\n",
    "class Pssm:\n",
    "    def __init__(self, filename):\n",
    "        self.load(filename)\n",
    "    \n",
    "    def load(self, filename):\n",
    "        sequences = []\n",
    "        \n",
    "        with open(filename) as file:\n",
    "            for line in file:\n",
    "                line = line.strip()\n",
    "                if not line.startswith(\">\"):\n",
    "                    sequences.append(line.upper())\n",
    "        \n",
    "        self.numberColumns = len(sequences[0])\n",
    "        matrix = [{aa : 0 for aa in aminoAcids} for _ in range(self.numberColumns)]\n",
    "        beta = len(sequences) ** 0.5\n",
    "        self.matrix = deepcopy(matrix)\n",
    "        \n",
    "        for sequence in sequences:\n",
    "            for i, aa in enumerate(sequence):\n",
    "                if aa in aminoAcids:\n",
    "                    matrix[i][aa] += 1\n",
    "                    \n",
    "        \n",
    "        for i, column in enumerate(matrix):\n",
    "            alpha = sum(column.values())\n",
    "            for aa, count in column.items():\n",
    "                frequency = count / len(sequences)\n",
    "                probability = (alpha * frequency + beta * p[aa]) / (alpha + beta)\n",
    "                self.matrix[i][aa] = log(probability / p[aa], logBase)\n",
    "    \n",
    "    def __getitem__(self, column, aa):\n",
    "        return self.matrix[column][aa]\n",
    "    \n",
    "    def __str__(self):\n",
    "        res = \"  | \"\n",
    "        cellWidth = 5\n",
    "        for i in range(self.numberColumns):\n",
    "            res += str(i).rjust(cellWidth) + \" | \"\n",
    "        res += \"\\n  \" + \"+-------\" * self.numberColumns + \"\\n\"\n",
    "        for aa in aminoAcids:\n",
    "            res += aa + \" | \"\n",
    "            for column in self.matrix:\n",
    "                res += str(column[aa])[:cellWidth].rjust(cellWidth) + \" | \"\n",
    "            res += \"\\n\"\n",
    "        res += \"\\n\"\n",
    "        return res\n",
    "\n",
    "def main():\n",
    "    muscle = \"msaresults-MUSCLE.fasta\"\n",
    "    clustal = \"msaresults-CLUSTAL.fasta\"\n",
    "    pssm = Pssm(muscle)\n",
    "    #pssm = Pssm(clustal)\n",
    "    print(pssm)\n",
    "    \n",
    "if __name__ == \"__main__\":\n",
    "    main()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Results analysis\n",
    "The constructed PSSMs can be found in `MUSCLE-PSSM.txt` and `CLUSTAL-PSSM.txt`, they aren't shown here because of their excessive width. We will compare the PSSM files and the weblogo constructed from the alignment fasta files.\n",
    "\n",
    "Of course these comparisons are not formal and do not prove that this project works for all domain alignments. The aim is just to convince that it works for this case, and it is sufficient, because this project only targets the WW domain.\n",
    "\n",
    "### MUSCLE Weblogo\n",
    "![MUSCLE-logo](MUSCLE-logo.png)\n",
    "By searching for high values on the weblogo and in `MUSCLE-PSSM.txt`, we can see that they all match. For exemple:\n",
    "* A big W on the weblogo at index 5, and a value of 6.407 for W in the PSSM at this index\n",
    "* D slightly bigger than T at index 10, and 3.013 for D and 2.333 for T in the PSSM  at this index\n",
    "* A big P on the weblogo at index 57, and a value of 4.290 for P in the PSSM at this index\n",
    "### CLUSTAL Weblogo\n",
    "![CLUSTAL-logo](CLUSTAL-logo.png)\n",
    "By searching for high values on the weblogo and in `CLUSTAL-PSSM.txt`, we can see that they all match. For exemple:\n",
    "* A big W on the weblogo at index 5, and a value of 6.407 for W in the PSSM at this index\n",
    "* A big G on the weblogo at index 14 and a value of 3.455 for G in the PSSM at this index\n",
    "* A big W on the weblogo at index 53 and a value of 6.127 for W in the PSSM at this index\n",
    "### Preserved positions in the Weblogos\n",
    "There are a few amino acids that are very well conserved in all sequences of the domain:\n",
    "* Amino acid P at index 2\n",
    "* Amino acid W at index 5\n",
    "* Amino acid Y at index 17/18\n",
    "* Amino acid Y at index 18/19\n",
    "* Amino acid H at index 21/47\n",
    "* Amino acid W at index 53/54\n",
    "* Amino acid P at index 56/57\n",
    "\n",
    "Note that the most prominent amino acids can be found in both alignment, but their exact indices slightly differ. This is due to different gaps in the alignments. The most noticeable difference is for an amino acid H that is before a big gap in the CLUSTAL alignment, and right after the gap in the MUSCLE alignment.\n",
    "### HMM logo\n",
    "Here is the HMM logo for the WW family:  \n",
    "\n",
    "![HMM-logo](HMM-logo.png)\n",
    "\n",
    "We can see that the preserved positions are the same as in the weblogos, although scaled differently. Gaps are shown in the HMM logo with vertical red lines, and by the last two lines of numbers below the image. The gap information in the HMM logo roughly matches with the observed gaps in the Weblogos."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2+"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
