{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from enum import Enum\n",
    "\n",
    "AminoAcid = Enum(\"AminoAcid\", \"A R N D C E Q G H I L K M F P S T W Y V B Z X\")\n",
    "\n",
    "class Sequence:\n",
    "    \"\"\"Represents a sequence of amino acids (with no further semantic).\n",
    "    It is suited to compute the best alignment between two sequences.\"\"\"\n",
    "    \n",
    "    GAP = None\n",
    "    def __init__(self, filename, sequenceId):\n",
    "        \"\"\"Constructor. Loads the sequence from a file.\n",
    "        \n",
    "        Parameters:\n",
    "            filename: the filename of the file containing the sequence.\n",
    "            sequenceId: the identifier of the sequence in the file.\n",
    "                This is the string between the two first bars \"|\" on the line preceding the sequence.\n",
    "            \"\"\"\n",
    "        self.loadFromFile(filename, sequenceId)\n",
    "    \n",
    "    def insertAminoAcid(self, aminoAcid, index):\n",
    "        \"\"\"Adds an amino acid to the sequence at the specific index.\"\"\"\n",
    "        self.sequence.insert(index, aminoAcid)\n",
    "    \n",
    "    def insertGap(self, index):\n",
    "        \"\"\"Inserts a gap in the sequence at the specified index.\"\"\"\n",
    "        self.sequence.insert(index, self.GAP)\n",
    "    \n",
    "    def getElement(self, index):\n",
    "        \"\"\"Returns the elements at the specified index.\"\"\"\n",
    "        return self.sequence[index]\n",
    "\n",
    "    def loadFromFile(self, filename, sequenceId):\n",
    "        \"\"\"Loads the sequence data from a file.\n",
    "        \n",
    "        Parameters:\n",
    "            filename: the filename of the file containing the sequence.\n",
    "            sequenceId: the identifier of the sequence in the file.\n",
    "                This is the string between the two first bars \"|\" on the line preceding the sequence.\n",
    "        \"\"\"\n",
    "        found = False\n",
    "        with open(filename) as file:\n",
    "            for line in file:\n",
    "                # If we found the needed identifier line\n",
    "                if line[0] == \">\" and line.split(\"|\")[1] == sequenceId:\n",
    "                    found = True\n",
    "                    sequence = file.readline()\n",
    "                    break\n",
    "\n",
    "        if not found:\n",
    "            raise RuntimeError(\"Sequence not found in the specified file.\")\n",
    "\n",
    "        try:\n",
    "            sequence = [AminoAcid[character] for character in sequence if not character.isspace()]\n",
    "        except KeyError as e:\n",
    "            raise RuntimeError(\"An amino acid is not valid in the sequence: \" + str(e))\n",
    "\n",
    "        # We didn't worked on self.sequence directly to preserve the internal state in case of failure\n",
    "        self.sequence = sequence\n",
    "    \n",
    "class Score:\n",
    "    \"\"\"Class used to compute the score of substitution between two amino acids, by using well known matrix\n",
    "    such as the BLOSUM.\"\"\"\n",
    "    \n",
    "    def __init__(self, filename):\n",
    "        \"\"\"Contructor. Loads the matrix from a file.\n",
    "        \n",
    "        Parameter:\n",
    "            filename: the filename of the matrix to load. It must be a .iij file as given in the blosum archive.\n",
    "        \"\"\"\n",
    "        self.loadFromFile(filename)\n",
    "    \n",
    "    def getScore(self, aminoAcidA, aminoAcidB):\n",
    "        \"\"\"Returns the score of matching between two amino acids according to the loaded matrix.\"\"\"\n",
    "        indexA = self.indices.index(aminoAcidA)\n",
    "        indexB = self.indices.index(aminoAcidB)\n",
    "        \n",
    "        # The matrix is triangular, so we need to test the two possibilities for indexing\n",
    "        # since one of the two would result in an index error\n",
    "        if len(self.matrix[indexA]) > indexB:\n",
    "            return self.matrix[indexA][indexB]\n",
    "        else:\n",
    "            return self.matrix[indexB][indexA]\n",
    "    \n",
    "    def loadFromFile(self, filename):\n",
    "        \"\"\"Loads a substitution matrix from a file.\n",
    "        \n",
    "        Parameter:\n",
    "            filename: the filename of the matrix to load. It must be a .iij file as given in the blosum archive.\n",
    "        \"\"\"\n",
    "        self.matrix = []\n",
    "        self.indeices = []\n",
    "        \n",
    "        with open(filename) as file:\n",
    "            foundAminoAcidList = False\n",
    "            for line in file:\n",
    "                # Skip commentary lines\n",
    "                if line.strip().startswith(\"#\"):\n",
    "                    continue\n",
    "                \n",
    "                if not foundAminoAcidList:\n",
    "                    # The first non-commentary line is the list of amino acids\n",
    "                    # We reload this list because some files may order the acids differently\n",
    "                    self.indices = [AminoAcid[character] for character in line.split()]\n",
    "                    foundAminoAcidList = True\n",
    "                else:\n",
    "                    self.matrix.append([float(number) for number in line.split()])\n",
    "                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "RuntimeError",
     "evalue": "An amino acid is not valid in the sequence: '\\n'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-3-1045a90a3a1f>\u001b[0m in \u001b[0;36mloadFromFile\u001b[0;34m(self, filename, sequenceId)\u001b[0m\n\u001b[1;32m     52\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 53\u001b[0;31m             \u001b[0msequence\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mAminoAcid\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcharacter\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mcharacter\u001b[0m \u001b[0;32min\u001b[0m \u001b[0msequence\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     54\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-3-1045a90a3a1f>\u001b[0m in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m     52\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 53\u001b[0;31m             \u001b[0msequence\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mAminoAcid\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcharacter\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mcharacter\u001b[0m \u001b[0;32min\u001b[0m \u001b[0msequence\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     54\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/lib/python3.5/enum.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(cls, name)\u001b[0m\n\u001b[1;32m    276\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__getitem__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcls\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 277\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mcls\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_member_map_\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    278\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: '\\n'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-6-4a881f593449>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mscore\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mScore\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"blosum/blosum40.iij\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0msequence\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mSequence\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"SH3-sequence.fasta\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"Q8IZP0\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msequence\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msequence\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-3-1045a90a3a1f>\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, filename, sequenceId)\u001b[0m\n\u001b[1;32m     16\u001b[0m                 \u001b[0mThis\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0mthe\u001b[0m \u001b[0mstring\u001b[0m \u001b[0mbetween\u001b[0m \u001b[0mthe\u001b[0m \u001b[0mtwo\u001b[0m \u001b[0mfirst\u001b[0m \u001b[0mbars\u001b[0m \u001b[0;34m\"|\"\u001b[0m \u001b[0mon\u001b[0m \u001b[0mthe\u001b[0m \u001b[0mline\u001b[0m \u001b[0mpreceding\u001b[0m \u001b[0mthe\u001b[0m \u001b[0msequence\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     17\u001b[0m             \"\"\"\n\u001b[0;32m---> 18\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloadFromFile\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilename\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msequenceId\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     19\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     20\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0minsertAminoAcid\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maminoAcid\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-3-1045a90a3a1f>\u001b[0m in \u001b[0;36mloadFromFile\u001b[0;34m(self, filename, sequenceId)\u001b[0m\n\u001b[1;32m     53\u001b[0m             \u001b[0msequence\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mAminoAcid\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcharacter\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mcharacter\u001b[0m \u001b[0;32min\u001b[0m \u001b[0msequence\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     54\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 55\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mRuntimeError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"An amino acid is not valid in the sequence: \"\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0me\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     56\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     57\u001b[0m         \u001b[0;31m# We didn't worked on self.sequence directly to preserve the internal state in case of failure\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mRuntimeError\u001b[0m: An amino acid is not valid in the sequence: '\\n'"
     ]
    }
   ],
   "source": [
    "score = Score(\"blosum/blosum40.iij\")\n",
    "sequence = Sequence(\"SH3-sequence.fasta\", \"Q8IZP0\")\n",
    "print(sequence.sequence)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
